// @firecms/firebase/dist/index.es.js
import we, { useState as L, useRef as xe, useCallback as E, useEffect as U } from "react";
import ct from "react-fast-compare";
import { getAuth as te, onAuthStateChanged as at, GoogleAuthProvider as Ge, signInWithPopup as Ne, signInWithEmailAndPassword as lt, createUserWithEmailAndPassword as ut, fetchSignInMethodsForEmail as dt, signOut as ft, signInAnonymously as ht, signInWithPhoneNumber as gt, OAuthProvider as Ie, FacebookAuthProvider as mt, GithubAuthProvider as pt, TwitterAuthProvider as wt, RecaptchaVerifier as Ye, getMultiFactorResolver as vt, PhoneMultiFactorGenerator as Te, PhoneAuthProvider as Oe } from "firebase/auth";
import { getStorage as ve, uploadBytes as yt, ref as ye, getDownloadURL as Re, getMetadata as Et } from "firebase/storage";
import { getApps as bt, deleteApp as Ct, initializeApp as Ft } from "firebase/app";
import { getToken as Pt, initializeAppCheck as St } from "firebase/app-check";
import { stripCollectionPath as Ee, COLLECTION_PATH_SEPARATOR as J, makePropertiesEditable as xt, sortProperties as kt, removeFunctions as Lt, removePropsIfExisting as Dt, GeoPoint as ke, EntityReference as jt, useModeController as Nt, ErrorView as Z, FireCMSLogo as It, useBrowserTitleAndIcon as Tt, useBuildModeController as Ot, useBuildLocalConfigurationPersistence as Rt, useValidateAuthenticator as Mt, useBuildNavigationController as _t, CircularProgressCenter as Me, SnackbarProvider as Ut, ModeControllerProvider as Ht, FireCMS as Vt, Scaffold as Bt, NavigationRoutes as $t, SideDialogs as zt, removeInitialAndTrailingSlashes as Je, unslugify as qt } from "@firecms/core";
import { deleteField as Le, getFirestore as O, query as $, collection as _, limit as W, getDocs as G, onSnapshot as Ce, collectionGroup as We, where as _e, orderBy as Gt, startAfter as Yt, getDoc as Jt, doc as q, setDoc as Wt, deleteDoc as At, getCountFromServer as Kt, serverTimestamp as Ae, Timestamp as re, GeoPoint as Ke, DocumentReference as Qe } from "firebase/firestore";
import * as Qt from "js-search";
import { getDatabase as V, query as ce, ref as M, orderByKey as Zt, startAt as Ue, limitToFirst as He, get as be, onValue as Ve, push as Be, set as Xt, remove as er, orderByChild as tr } from "firebase/database";
import { jsx as i, Fragment as oe, jsxs as D } from "react/jsx-runtime";
import { EmailIcon as rr, PhoneIcon as or, PersonOutlineIcon as nr, Button as A, IconButton as Ze, ArrowBackIcon as Xe, Typography as X, TextField as le, CircularProgress as et, CenteredView as ir } from "@firecms/ui";
import { buildEntityPropertiesFromData as tt, buildPropertiesOrder as sr } from "@firecms/schema_inference";
const cr = ({
  loading: e,
  firebaseApp: t,
  signInOptions: n,
  onSignOut: r,
  defineRolesFor: s
}) => {
  const [l, c] = L(void 0), [d, v] = L(), [a, o] = L(), [u, g] = L(!0), [h, f] = L(!0), [w, b] = L(!1), [P, C] = L(), [F, y] = L(), [x, I] = L(), j = xe(t ? te(t) : null), H = E(async (m, p) => {
    console.log("Updating user", m), !e && (c(m), f(!1), p && g(!1));
  }, [e]), z = E(async (m) => {
    if (s && m) {
      const p = await s(m);
      ct(p, F) || y(p);
    }
  }, [s, F]);
  U(() => {
    z && l && z(l);
  }, [z, l]), U(() => {
    if (t)
      try {
        const m = te(t);
        return j.current = m, v(void 0), H(m.currentUser, !1), at(
          m,
          async (p) => {
            console.log("User state changed", p), await H(p, !0);
          },
          (p) => o(p)
        );
      } catch (m) {
        return v(m), g(!1), () => {
        };
      }
  }, [t, H]), U(() => {
    !e && j.current && H(j.current.currentUser, !1);
  }, [e, H]);
  const k = E((m) => n?.find((p) => {
    if (p === null)
      throw Error("useFirebaseAuthController");
    if (typeof p == "object" && p.provider === m)
      return p;
  }), [n]), K = E(() => {
    const m = new Ge(), p = k("google.com");
    p?.scopes && p.scopes.forEach((T) => m.addScope(T)), p?.customParameters ? m.setCustomParameters(p.customParameters) : m.setCustomParameters({
      prompt: "select_account"
    });
    const S = j.current;
    if (!S)
      throw Error("No auth");
    Ne(S, m).catch(o);
  }, [k]), Y = E(async () => {
    if (!l)
      throw Error("No client user is logged in");
    return l.getIdToken();
  }, [l]), ne = E((m, p) => {
    const S = j.current;
    if (!S)
      throw Error("No auth");
    f(!0), lt(S, m, p).catch(o).then(() => f(!1));
  }, []), de = E((m, p) => {
    const S = j.current;
    if (!S)
      throw Error("No auth");
    f(!0), ut(S, m, p).catch(o).then(() => f(!1));
  }, []), De = E((m) => {
    const p = j.current;
    if (!p)
      throw Error("No auth");
    return f(!0), dt(p, m).then((S) => (f(!1), S));
  }, []), je = E(() => {
    const m = j.current;
    if (!m)
      throw Error("No auth");
    ft(m).then((p) => {
      c(null), y(void 0), o(null), r && r();
    }), b(!1);
  }, [r]), N = E((m, p) => {
    f(!0), Ne(m, p).catch(o).then(() => f(!1));
  }, []), fe = E(() => {
    const m = j.current;
    if (!m)
      throw Error("No auth");
    f(!0), ht(m).catch(o).then(() => f(!1));
  }, []), Q = E((m, p) => {
    const S = j.current;
    if (!S)
      throw Error("No auth");
    return f(!0), gt(S, m, p).catch(o).then((T) => {
      f(!1), C(T ?? void 0);
    });
  }, []), ie = E(() => {
    const m = new Ie("apple.com"), p = k("apple.com");
    p?.scopes && p.scopes.forEach((T) => m.addScope(T)), p?.customParameters && m.setCustomParameters(p.customParameters);
    const S = j.current;
    if (!S)
      throw Error("No auth");
    N(S, m);
  }, [N, k]), he = E(() => {
    const m = new mt(), p = k("facebook.com");
    p?.scopes && p.scopes.forEach((T) => m.addScope(T)), p?.customParameters && m.setCustomParameters(p.customParameters);
    const S = j.current;
    if (!S)
      throw Error("No auth");
    N(S, m);
  }, [N, k]), ge = E(() => {
    const m = new pt(), p = k("github.com");
    p?.scopes && p.scopes.forEach((T) => m.addScope(T)), p?.customParameters && m.setCustomParameters(p.customParameters);
    const S = j.current;
    if (!S)
      throw Error("No auth");
    N(S, m);
  }, [N, k]), me = E(() => {
    const m = new Ie("microsoft.com"), p = k("microsoft.com");
    p?.scopes && p.scopes.forEach((T) => m.addScope(T)), p?.customParameters && m.setCustomParameters(p.customParameters);
    const S = j.current;
    if (!S)
      throw Error("No auth");
    N(S, m);
  }, [N, k]), pe = E(() => {
    const m = new wt(), p = k("twitter.com");
    p?.customParameters && m.setCustomParameters(p.customParameters);
    const S = j.current;
    if (!S)
      throw Error("No auth");
    N(S, m);
  }, [N, k]), R = E(() => {
    b(!0), c(null), y(void 0);
  }, []);
  return {
    user: l ? {
      ...l,
      roles: F,
      firebaseUser: l
    } : null,
    roles: F,
    setUser: H,
    setRoles: y,
    authProviderError: a,
    authLoading: h,
    initialLoading: e || u,
    signOut: je,
    getAuthToken: Y,
    googleLogin: K,
    skipLogin: R,
    loginSkipped: w,
    emailPasswordLogin: ne,
    createUserWithEmailAndPassword: de,
    fetchSignInMethodsForEmail: De,
    anonymousLogin: fe,
    phoneLogin: Q,
    appleLogin: ie,
    facebookLogin: he,
    githubLogin: ge,
    microsoftLogin: me,
    twitterLogin: pe,
    confirmationResult: P,
    extra: x,
    setExtra: I
  };
};
function ar({ firebaseApp: e, bucketUrl: t }) {
  const n = {};
  return {
    uploadFile({ file: r, fileName: s, path: l, metadata: c, bucket: d }) {
      if (!e)
        throw Error("useFirebaseStorageSource Firebase not initialised");
      const a = ve(e, d ?? t);
      if (!a)
        throw Error("useFirebaseStorageSource Firebase not initialised");
      const o = s ?? r.name;
      return console.debug("Uploading file", o, r, l, c), yt(ye(a, `${l}/${o}`), r, c).then((u) => ({
        path: u.ref.fullPath
      }));
    },
    async getFile(r, s) {
      try {
        if (!e)
          throw Error("useFirebaseStorageSource Firebase not initialised");
        const c = ve(e, s ?? t);
        if (!c)
          throw Error("useFirebaseStorageSource Firebase not initialised");
        const d = ye(c, r), v = await Re(d), o = await (await fetch(v)).blob();
        return new File([o], r);
      } catch (l) {
        if (l?.code === "storage/object-not-found")
          return null;
        throw l;
      }
    },
    async getDownloadURL(r, s) {
      if (!e)
        throw Error("useFirebaseStorageSource Firebase not initialised");
      const c = ve(e, s ?? t);
      if (!c)
        throw Error("useFirebaseStorageSource Firebase not initialised");
      if (n[r])
        return n[r];
      try {
        const d = ye(c, r), [v, a] = await Promise.all([Re(d), Et(d)]), o = {
          url: v,
          metadata: a
        };
        return n[r] = o, o;
      } catch (d) {
        if (d?.code === "storage/object-not-found")
          return { url: null, fileNotFound: !0 };
        throw d;
      }
    }
  };
}
const lr = `It seems like the provided Firebase config is not correct. If you 
are using the credentials provided automatically by Firebase 
Hosting, make sure you link your Firebase app to Firebase Hosting. 
`;
function ur({
  firebaseConfig: e,
  fromUrl: t,
  onFirebaseInit: n,
  name: r,
  authDomain: s
}) {
  const [l, c] = L(), [d, v] = L(!1), [a, o] = L(), u = E((g) => {
    try {
      const h = r ?? "[DEFAULT]", w = bt().find((P) => P.name === h);
      w && Ct(w);
      const b = Ft(g, h);
      o(void 0), v(!1), c(b);
    } catch (h) {
      console.error("Error initialising Firebase", h), o(lr + `
` + (h.message ?? JSON.stringify(h)));
    }
  }, [r]);
  return U(() => {
    n && e && l && n(e, l);
  }, [l]), U(() => {
    v(!0);
    function g(h) {
      fetch(h).then(async (f) => {
        if (console.debug("Firebase init response", f.status), f && f.status < 300) {
          const w = await f.json();
          s && (w.authDomain = s), u(w);
        }
      }).catch(
        (f) => {
          v(!1), o(
            "Could not load Firebase configuration from Firebase hosting. If the app is not deployed in Firebase hosting, you need to specify the configuration manually" + f.toString()
          );
        }
      );
    }
    e ? u(e) : t ? g(t) : process.env.NODE_ENV === "production" ? g("/__/firebase/init.json") : (v(!1), o(
      "You need to deploy the app to Firebase hosting or specify a Firebase configuration object"
    ));
  }, []), {
    firebaseApp: l,
    firebaseConfigLoading: d,
    configError: a
  };
}
function dr({
  firebaseApp: e,
  options: t
}) {
  t?.debugToken && Object.assign(window, {
    FIREBASE_APPCHECK_DEBUG_TOKEN: t?.debugToken
  });
  const [n, r] = we.useState(!1), [s, l] = we.useState(void 0), [c, d] = we.useState(), v = E(async (a) => {
    try {
      const o = await Pt(a, t?.forceRefresh);
      console.debug("App Check token:", o), o ? (l(!0), console.debug("App Check success.")) : (d("App Check failed."), l(!1));
    } catch (o) {
      console.error(o), d(o.message);
    }
  }, [t?.forceRefresh]);
  return U(() => {
    if (!t || !e || s !== void 0)
      return;
    r(!0);
    const {
      provider: a,
      isTokenAutoRefreshEnabled: o
    } = t, u = St(e, {
      provider: a,
      isTokenAutoRefreshEnabled: o
    });
    v(u).then(() => {
      r(!1);
    });
  }, [s, e, t, v]), {
    loading: n,
    appCheckVerified: s,
    error: c
  };
}
function Yr(e, t) {
  return t ? [...t.map(Ee), Ee(e)].join(J) : Ee(e);
}
function Fe(e) {
  return Array.isArray(e) ? e.map((t) => Fe(t)) : typeof e == "object" ? Object.entries(e).map(([t, n]) => ({ [t]: Fe(n) })).reduce((t, n) => ({ ...t, ...n }), {}) : e === void 0 ? Le() : e;
}
const Jr = (e) => {
  const t = e.map((r) => {
    const s = r.id, l = fr(r);
    return { [s]: l };
  }).reduce((r, s) => ({ ...r, ...s }), {});
  return Object.keys(t).sort((r, s) => s.split(J).length - r.split(J).length).forEach((r) => {
    const s = t[r];
    if (r.includes(J)) {
      const l = r.split(J).slice(0, -1).join(J), c = t[l];
      c && (c.subcollections = [...c.subcollections ?? [], s]), delete t[r];
    }
  }), Object.values(t);
}, fr = (e) => {
  const t = e.data();
  if (!t)
    throw Error("Entity collection has not been persisted correctly");
  const n = t.propertiesOrder, r = t.properties ?? {};
  xt(r);
  const s = kt(r, n);
  return {
    ...t,
    properties: s,
    id: t.id ?? t.alias ?? t.path
  };
};
function Wr(e, t) {
  const { properties: n, ...r } = e, s = n ? hr(n, t) : void 0, l = s ? Fe(Lt(s)) : void 0;
  let c = {};
  return r && (c = {
    ...r
  }), l && (c.properties = l), (r.propertiesOrder || l) && (c.propertiesOrder = gr(r.propertiesOrder ?? Object.keys(l))), delete c.permissions, c.entityViews && (c.entityViews = c.entityViews.filter((d) => typeof d == "string")), delete c.editable, delete c.additionalFields, delete c.callbacks, delete c.Actions, delete c.entityActions, delete c.selectionController, delete c.subcollections, delete c.exportable, c;
}
function hr(e, t) {
  const n = {};
  return Object.entries(e).forEach(([r, s]) => {
    if (typeof s == "object") {
      const l = s.propertyConfig ? t[s.propertyConfig] : void 0;
      let c;
      l?.property ? c = Dt(s, l?.property) : c = s, n[r] = { ...c };
    }
  }), n;
}
function gr(e) {
  return [...new Set(e)];
}
async function Ar(e, t, n, r) {
  console.debug("getFirestoreDataInPath", t, r);
  const s = O(e);
  if (n) {
    let l, c = 0;
    const d = n;
    d.push(t);
    let v = d[0];
    for (; v; ) {
      if (l)
        l = (await Promise.all(l.map(async (a) => {
          const o = $(_(s, a.ref.path, v), W(5));
          return (await G(o)).docs;
        }))).flat();
      else {
        const a = $(_(s, v), W(5));
        l = (await G(a)).docs;
      }
      c++, v = c < d.length ? d[c] : void 0;
    }
    return l ? l.map((a) => a.data()) : [];
  } else {
    const l = $(_(s, t), W(r));
    return G(l).then((c) => c.docs.map((d) => d.data()));
  }
}
function Kr(e, t) {
  return console.debug("Performing Algolia query", e, t), e.search(t).then(({ hits: n }) => n.map((r) => r.objectID)).catch((n) => (console.error(n), []));
}
function Qr({
  isPathSupported: e,
  search: t
}) {
  return (n) => ({
    init: (s) => Promise.resolve(e(s.path)),
    search: t
  });
}
const $e = "https://api-drplyi3b6q-ey.a.run.app";
async function Zr({
  host: e = $e,
  firebaseToken: t,
  projectId: n,
  collectionPath: r,
  query: s
}) {
  return console.debug("Performing Pinecone query", r, s), (await (await fetch(
    (e ?? $e) + `/projects/${n}/search/${r}`,
    {
      // mode: "no-cors",
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Basic ${t}`
        // "x-de-version": version
      },
      body: JSON.stringify({
        query: s
      })
    }
  )).json()).data.ids;
}
function Xr({
  isPathSupported: e,
  search: t
}) {
  return (n) => ({
    init: (s) => Promise.resolve(e(s.path)),
    search: t
  });
}
const mr = 100, ze = ({
  firebaseApp: e
}) => {
  let t;
  const n = {}, r = {}, s = (d) => {
    r[d] && (r[d](), delete r[d]);
  };
  return {
    init: ({
      path: d,
      collection: v
    }) => (console.debug("Init local search controller", d, v), t && d !== t && s(t), t = d, new Promise((a, o) => {
      if (!n[d] && v) {
        console.debug("Init local search controller", d), n[d] = pr(v);
        const u = O(e), g = _(u, d);
        console.log("Listening to collection", d, g), r[d] = Ce(
          $(g),
          {
            next: (h) => {
              const f = h.docs.map((w) => ({
                id: w.id,
                ...w.data()
              }));
              console.log("Added docs to index", d, f.length), n[d].addDocuments(f), a(!0);
            },
            error: (h) => {
              console.error("Error initializing local search controller", d), console.error(h), o(h);
            }
          }
        );
      }
    })),
    search: async ({ searchString: d, path: v }) => {
      const a = n[v];
      if (!a)
        throw new Error(`Index not found for path ${v}`);
      return (await a.search(d)).splice(0, mr).map((u) => u.id);
    }
  };
};
function pr(e) {
  const t = new Qt.Search("id");
  return t.addIndex("id"), Object.entries(e.properties).forEach(([n, r]) => {
    t.addIndex(n);
  }), t;
}
function wr({
  firebaseApp: e,
  textSearchControllerBuilder: t,
  firestoreIndexesBuilder: n,
  localTextSearchEnabled: r
}) {
  const s = xe();
  U(() => {
    !s.current && e && (t || r) && !s.current && (s.current = Er({
      firebaseApp: e,
      textSearchControllerBuilder: t,
      localTextSearchEnabled: r ?? !1
    }));
  }, [e, r, t]);
  const l = E((a, o, u, g, h, f, w = !1) => {
    if (!e)
      throw Error("useFirestoreDelegate Firebase not initialised");
    const b = O(e), P = w ? We(b, a) : _(b, a), C = [];
    return o && Object.entries(o).filter(([F, y]) => !!y).forEach(([F, y]) => {
      const [x, I] = y;
      C.push(_e(F, x, ee(I, b)));
    }), u && g && C.push(Gt(u, g)), h && C.push(Yt(h)), f && C.push(W(f)), $(P, ...C);
  }, [e]), c = E((a, o) => {
    if (!e)
      throw Error("useFirestoreDelegate Firebase not initialised");
    const u = O(e);
    return Jt(q(u, a, o)).then((g) => {
      if (g.exists())
        return ae(g);
    });
  }, [e]), d = E(({
    path: a,
    entityId: o,
    collection: u,
    onUpdate: g,
    onError: h
  }) => {
    if (!e)
      throw Error("useFirestoreDelegate Firebase not initialised");
    const f = O(e);
    return Ce(
      q(f, a, o),
      {
        next: (w) => {
          g(ae(w));
        },
        error: h
      }
    );
  }, [e]), v = E(({
    path: a,
    searchString: o,
    onUpdate: u
  }) => {
    if (!e)
      throw Error("useFirestoreDelegate Firebase not initialised");
    const g = s.current;
    if (!g)
      throw Error("Trying to make text search without specifying a FirestoreTextSearchController");
    let h = [];
    const w = te(e).currentUser;
    if (!w)
      throw Error("No current user");
    const b = g.search({
      path: a,
      searchString: o,
      currentUser: w
    });
    if (!b)
      throw Error("The current path is not supported by the specified FirestoreTextSearchController");
    return b.then((P) => {
      P.length === 0 && (h = [], u([]));
      const C = [], F = /* @__PURE__ */ new Set();
      h = P.map(
        (y) => d({
          path: a,
          entityId: y,
          onUpdate: (x) => {
            x.values ? F.has(x.id) || (F.add(x.id), C.push(x), u(C)) : (F.delete(x.id), u([...C.filter((I) => I.id !== y)]));
          }
        })
      );
    }), () => {
      h.forEach((P) => P());
    };
  }, [e, d]);
  return {
    setDateToMidnight: vr,
    delegateToCMSModel: ue,
    cmsToDelegateModel: (a) => {
      if (!e)
        throw Error("useFirestoreDelegate Firebase not initialised");
      return ee(a, O(e));
    },
    currentTime: yr,
    initTextSearchController: E(async (a) => s.current ? s.current.init(a) : !1, []),
    /**
     * Fetch entities in a Firestore path
     * @param path
     * @param collection
     * @param filter
     * @param limit
     * @param startAfter
     * @param searchString
     * @param orderBy
     * @param order
     * @return Function to cancel subscription
     * @see useCollectionFetch if you need this functionality implemented as a hook
     * @group Firestore
     */
    fetchCollection: E(async ({
      path: a,
      filter: o,
      limit: u,
      startAfter: g,
      searchString: h,
      orderBy: f,
      order: w,
      isCollectionGroup: b
    }) => {
      console.debug("Fetching collection", {
        path: a,
        limit: u,
        filter: o,
        startAfter: g,
        orderBy: f,
        order: w,
        isCollectionGroup: b
      });
      const P = l(a, o, f, w, g, u, b);
      return (await G(P)).docs.map((F) => ae(F));
    }, [l]),
    /**
     * Listen to a entities in a given path
     * @param path
     * @param collection
     * @param onError
     * @param filter
     * @param limit
     * @param startAfter
     * @param searchString
     * @param orderBy
     * @param order
     * @param onUpdate
     * @return Function to cancel subscription
     * @see useCollectionFetch if you need this functionality implemented as a hook
     * @group Firestore
     */
    listenCollection: E(({
      path: a,
      filter: o,
      limit: u,
      startAfter: g,
      searchString: h,
      orderBy: f,
      order: w,
      onUpdate: b,
      onError: P,
      isCollectionGroup: C,
      collection: F
    }) => {
      if (console.debug("Listening collection", {
        path: a,
        searchString: h,
        isCollectionGroup: C,
        limit: u,
        filter: o,
        startAfter: g,
        orderBy: f,
        order: w
      }), !e)
        throw Error("useFirestoreDelegate Firebase not initialised");
      const y = l(a, o, f, w, g, u, C);
      return h ? v({
        path: a,
        searchString: h,
        onUpdate: b
      }) : Ce(
        y,
        {
          next: (x) => {
            h || b(x.docs.map((I) => ae(I)));
          },
          error: P
        }
      );
    }, [l, e, v]),
    /**
     * Retrieve an entity given a path and a collection
     * @param path
     * @param entityId
     * @param collection
     * @group Firestore
     */
    fetchEntity: E(({
      path: a,
      entityId: o
    }) => c(a, o), [c]),
    /**
     *
     * @param path
     * @param entityId
     * @param collection
     * @param onUpdate
     * @param onError
     * @return Function to cancel subscription
     * @group Firestore
     */
    listenEntity: d,
    /**
     * Save entity to the specified path. Note that Firestore does not allow
     * undefined values.
     * @param path
     * @param entityId
     * @param values
     * @param schemaId
     * @param collection
     * @param status
     * @group Firestore
     */
    saveEntity: E(({
      path: a,
      entityId: o,
      values: u,
      status: g
    }) => {
      if (!e)
        throw Error("useFirestoreDelegate Firebase not initialised");
      const h = O(e), f = _(h, a);
      console.debug("Saving entity", {
        path: a,
        entityId: o,
        values: u
        // cleanedValues
      });
      let w;
      return o ? w = q(f, o) : w = q(f), Wt(w, u, { merge: !0 }).then(() => ({
        id: w.id,
        path: a,
        values: u
      })).catch((b) => {
        throw console.error("Error saving entity", b), b;
      });
    }, [e]),
    /**
     * Delete an entity
     * @param entity
     * @param collection
     * @group Firestore
     */
    deleteEntity: E(({
      entity: a
    }) => {
      if (!e)
        throw Error("useFirestoreDelegate Firebase not initialised");
      const o = O(e);
      return At(q(o, a.path, a.id));
    }, [e]),
    /**
     * Check if the given property is unique in the given collection
     * @param path Collection path
     * @param name of the property
     * @param value
     * @param property
     * @param entityId
     * @return `true` if there are no other fields besides the given entity
     * @group Firestore
     */
    checkUniqueField: E(async (a, o, u, g) => {
      if (!e)
        throw Error("useFirestoreDelegate Firebase not initialised");
      const h = O(e);
      if (console.debug("Check unique field entity", a, o, u, g), u == null)
        return Promise.resolve(!0);
      const f = $(_(h, a), _e(o, "==", ee(u, h)));
      return (await G(f)).docs.filter((b) => b.id !== g).length === 0;
    }, [e]),
    generateEntityId: E((a) => {
      if (!e)
        throw Error("useFirestoreDelegate Firebase not initialised");
      const o = O(e);
      return q(_(o, a)).id;
    }, [e]),
    countEntities: E(async ({
      path: a,
      filter: o,
      order: u,
      orderBy: g,
      isCollectionGroup: h
    }) => {
      if (!e)
        throw Error("useFirestoreDelegate Firebase not initialised");
      const f = l(a, o, g, u, void 0, void 0, h);
      return (await Kt(f)).data().count;
    }, [e]),
    isFilterCombinationValid: E(({
      path: a,
      collection: o,
      filterValues: u,
      sortBy: g
    }) => {
      if (!e)
        throw Error("useFirestoreDelegate Firebase not initialised");
      if (n === void 0)
        return !0;
      const h = n?.({
        path: a,
        collection: o
      }), f = g ? g[0] : void 0, w = g ? g[1] : void 0, b = Object.values(u), P = Object.keys(u), C = P.length;
      return !f && b.every((F) => F[0] === "==") || C === 1 && (!f || f === P[0]) ? !0 : !h && C > 1 ? !1 : !!h && h.filter((F) => !f || f in F).find(
        (F) => Object.entries(u).every(([y, x]) => F[y] !== void 0 && (!w || F[y] === w))
      ) !== void 0;
    }, [e])
  };
}
const ae = (e) => {
  const t = ue(e.data());
  return {
    id: e.id,
    path: rt(e.ref.path),
    values: t
  };
};
function ue(e) {
  if (e == null)
    return null;
  if (!Le().isEqual(e)) {
    if (Ae().isEqual(e))
      return null;
    if (e instanceof re || typeof e.toDate == "function" && e.toDate() instanceof Date)
      return e.toDate();
    if (e instanceof Date)
      return e;
    if (!(typeof e == "object" && "__type__" in e && e.__type__ === "__vector__")) {
      if (e instanceof Ke)
        return new ke(e.latitude, e.longitude);
      if (e instanceof Qe)
        return new jt(e.id, rt(e.path));
      if (Array.isArray(e))
        return e.map(ue).filter((t) => t !== void 0);
      if (typeof e == "object") {
        const t = {};
        for (const n of Object.keys(e)) {
          const r = ue(e[n]);
          r !== void 0 && (t[n] = r);
        }
        return t;
      }
      return e;
    }
  }
}
function rt(e) {
  let t = e.lastIndexOf("/");
  return t = t === -1 ? e.length : t, e.substring(0, t);
}
function vr(e) {
  if (!e || !(e instanceof re))
    return e;
  const t = e.toDate();
  return t.setHours(0, 0, 0, 0), re.fromDate(t);
}
function ee(e, t) {
  return e === void 0 ? Le() : e === null ? null : Array.isArray(e) ? e.filter((n) => n !== void 0).map((n) => ee(n, t)) : e.isEntityReference && e.isEntityReference() ? q(t, e.path, e.id) : e instanceof ke ? new Ke(e.latitude, e.longitude) : e instanceof Date ? re.fromDate(e) : e && typeof e == "object" && "__type__" in e && e.__type__ === "__vector__" ? void 0 : e && typeof e == "object" ? Object.entries(e).map(([n, r]) => {
    const s = ee(r, t);
    return s !== void 0 ? { [n]: s } : {};
  }).reduce((n, r) => ({ ...n, ...r }), {}) : e;
}
function yr() {
  return Ae();
}
function Er({
  textSearchControllerBuilder: e,
  firebaseApp: t,
  localTextSearchEnabled: n
}) {
  if (!e && n)
    return ze({ firebaseApp: t });
  if (!n && e)
    return e({ firebaseApp: t });
  if (!e && !n)
    return;
  const r = ze({ firebaseApp: t }), s = e({ firebaseApp: t });
  return {
    init: async (l) => await s.init(l) ? !0 : r.init(l),
    search: (l) => s.search(l) ?? r.search(l)
  };
}
function eo({ firebaseApp: e }) {
  const t = E(async ({
    path: o,
    filter: u,
    limit: g,
    startAfter: h,
    orderBy: f,
    order: w,
    searchString: b
  }) => {
    if (!e)
      throw new Error("Firebase app not provided");
    const P = V(e);
    let C = ce(M(P, o));
    h !== void 0 && (C = ce(C, Zt(), Ue(h))), g !== void 0 && (C = ce(C, He(g)));
    const F = await be(C);
    return F.exists() ? Object.entries(F.val()).map(([y, x]) => ({
      id: y,
      path: o,
      values: x
    })) : [];
  }, [e]), n = E(({
    path: o,
    onUpdate: u
    // Realtime Database does not directly support onError in onValue
  }) => {
    if (!e)
      throw new Error("Firebase app not provided");
    const g = V(e), h = M(g, o), f = Ve(h, (w) => {
      if (w.exists()) {
        const b = Object.entries(w.val()).map(([P, C]) => ({
          id: P,
          path: o,
          values: C
        }));
        u(b);
      } else
        u([]);
    });
    return () => f();
  }, [e]), r = E(async ({
    path: o,
    entityId: u
  }) => {
    if (!e)
      throw new Error("Firebase app not provided");
    const g = V(e), h = await be(M(g, `${o}/${u}`));
    if (h.exists())
      return {
        id: u,
        path: o,
        values: h.val()
      };
  }, [e]), s = E(({
    path: o,
    entityId: u,
    onUpdate: g,
    onError: h
  }) => {
    if (!e)
      throw new Error("Firebase app not provided");
    const f = V(e), w = M(f, `${o}/${u}`), b = Ve(w, (P) => {
      P.exists() ? g({
        id: u,
        path: o,
        values: P.val()
      }) : h?.(new Error("Entity does not exist"));
    });
    return () => b();
  }, [e]), l = E(async ({
    path: o,
    entityId: u,
    values: g
  }) => {
    if (!e)
      throw new Error("Firebase app not provided");
    const h = V(e), f = u ?? Be(M(h, o)).key;
    if (!f)
      throw new Error("Could not generate a new id");
    return await Xt(M(h, `${o}/${f}`), g), {
      id: f,
      path: o,
      values: g
    };
  }, [e]), c = E(async ({
    entity: o
  }) => {
    if (!e)
      throw new Error("Firebase app not provided");
    const u = V(e);
    await er(M(u, `${o.path}/${o.id}`));
  }, [e]), d = E(async (o, u, g, h) => {
    if (!e)
      throw new Error("Firebase app not provided");
    const f = V(e), w = ce(M(f, o), tr(u), Ue(g), He(1)), b = await be(w);
    if (!b.exists())
      return !0;
    const [P, C] = Object.entries(b.val())[0];
    return !!(C && typeof C == "object" && C[u] === g && P === h);
  }, [e]), v = E((o) => {
    if (!e)
      throw new Error("Firebase app not provided");
    const u = V(e);
    return Be(M(u, o)).key;
  }, [e]), a = E(({
    path: o,
    filter: u,
    sortBy: g
  }) => !1, []);
  return {
    fetchCollection: t,
    listenCollection: n,
    fetchEntity: r,
    listenEntity: s,
    saveEntity: l,
    deleteEntity: c,
    checkUniqueField: d,
    generateEntityId: v,
    isFilterCombinationValid: a,
    cmsToDelegateModel: (o) => {
      if (!e)
        throw new Error("Firebase app not provided");
      const u = V(e);
      return Se(o, u);
    },
    currentTime: () => /* @__PURE__ */ new Date(),
    delegateToCMSModel: Pe,
    setDateToMidnight: br
  };
}
function br(e) {
  const t = new Date(e);
  return t.setHours(0, 0, 0, 0), t;
}
function Pe(e) {
  if (e == null)
    return null;
  if (Array.isArray(e))
    return e.map(Pe).filter((t) => t !== void 0);
  if (typeof e == "object") {
    const t = {};
    for (const n of Object.keys(e)) {
      const r = Pe(e[n]);
      r !== void 0 && (t[n] = r);
    }
    return t;
  }
  return e;
}
function Se(e, t) {
  return e === void 0 || e === null ? null : Array.isArray(e) ? e.filter((n) => n !== void 0).map((n) => Se(n, t)) : e.isEntityReference && e.isEntityReference() ? M(t, `${e.path}/${e.id}`) : e instanceof Date ? e.toISOString() : e && typeof e == "object" ? Object.entries(e).map(([n, r]) => {
    const s = Se(r, t);
    return s !== void 0 ? { [n]: s } : {};
  }).reduce((n, r) => ({ ...n, ...r }), {}) : e;
}
const ot = "recaptcha-container";
function Cr() {
  return U(() => {
    if (!window || window?.recaptchaVerifier)
      return;
    const e = te();
    window.recaptchaVerifier = new Ye(
      e,
      ot,
      {
        size: "invisible"
      }
    );
  }, []), null;
}
const Fr = (e) => /* @__PURE__ */ i(oe, { children: /* @__PURE__ */ D(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    viewBox: "0 0 64 64",
    width: 32,
    height: 32,
    children: [
      /* @__PURE__ */ D(
        "linearGradient",
        {
          id: "95yY7w43Oj6n2vH63j6HJb",
          x1: "29.401",
          x2: "29.401",
          y1: "4.064",
          y2: "106.734",
          gradientTransform: "matrix(1 0 0 -1 0 66)",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ i("stop", { offset: "0", stopColor: "#ff5840" }),
            /* @__PURE__ */ i("stop", { offset: ".007", stopColor: "#ff5840" }),
            /* @__PURE__ */ i("stop", { offset: ".989", stopColor: "#fa528c" }),
            /* @__PURE__ */ i("stop", { offset: "1", stopColor: "#fa528c" })
          ]
        }
      ),
      /* @__PURE__ */ i(
        "path",
        {
          fill: "url(#95yY7w43Oj6n2vH63j6HJb)",
          d: "M47.46,15.5l-1.37,1.48c-1.34,1.44-3.5,1.67-5.15,0.6c-2.71-1.75-6.43-3.13-11-2.37 c-4.94,0.83-9.17,3.85-11.64, 7.97l-8.03-6.08C14.99,9.82,23.2,5,32.5,5c5,0,9.94,1.56,14.27,4.46 C48.81,10.83,49.13,13.71,47.46,15.5z"
        }
      ),
      /* @__PURE__ */ D(
        "linearGradient",
        {
          id: "95yY7w43Oj6n2vH63j6HJc",
          x1: "12.148",
          x2: "12.148",
          y1: ".872",
          y2: "47.812",
          gradientTransform: "matrix(1 0 0 -1 0 66)",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ i("stop", { offset: "0", stopColor: "#feaa53" }),
            /* @__PURE__ */ i("stop", { offset: ".612", stopColor: "#ffcd49" }),
            /* @__PURE__ */ i("stop", { offset: "1", stopColor: "#ffde44" })
          ]
        }
      ),
      /* @__PURE__ */ i(
        "path",
        {
          fill: "url(#95yY7w43Oj6n2vH63j6HJc)",
          d: "M16.01,30.91c-0.09,2.47,0.37,4.83,1.27,6.96l-8.21,6.05c-1.35-2.51-2.3-5.28-2.75-8.22 c-1.06-6.88,0.54-13.38, 3.95-18.6l8.03,6.08C16.93,25.47,16.1,28.11,16.01,30.91z"
        }
      ),
      /* @__PURE__ */ D(
        "linearGradient",
        {
          id: "95yY7w43Oj6n2vH63j6HJd",
          x1: "29.76",
          x2: "29.76",
          y1: "32.149",
          y2: "-6.939",
          gradientTransform: "matrix(1 0 0 -1 0 66)",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ i("stop", { offset: "0", stopColor: "#42d778" }),
            /* @__PURE__ */ i("stop", { offset: ".428", stopColor: "#3dca76" }),
            /* @__PURE__ */ i("stop", { offset: "1", stopColor: "#34b171" })
          ]
        }
      ),
      /* @__PURE__ */ i(
        "path",
        {
          fill: "url(#95yY7w43Oj6n2vH63j6HJd)",
          d: "M50.45,51.28c-4.55,4.07-10.61,6.57-17.36,6.71C22.91,58.2,13.66,52.53,9.07,43.92l8.21-6.05 C19.78,43.81, 25.67,48,32.5,48c3.94,0,7.52-1.28,10.33-3.44L50.45,51.28z"
        }
      ),
      /* @__PURE__ */ D(
        "linearGradient",
        {
          id: "95yY7w43Oj6n2vH63j6HJe",
          x1: "46",
          x2: "46",
          y1: "3.638",
          y2: "35.593",
          gradientTransform: "matrix(1 0 0 -1 0 66)",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ i("stop", { offset: "0", stopColor: "#155cde" }),
            /* @__PURE__ */ i("stop", { offset: ".278", stopColor: "#1f7fe5" }),
            /* @__PURE__ */ i("stop", { offset: ".569", stopColor: "#279ceb" }),
            /* @__PURE__ */ i("stop", { offset: ".82", stopColor: "#2cafef" }),
            /* @__PURE__ */ i("stop", { offset: "1", stopColor: "#2eb5f0" })
          ]
        }
      ),
      /* @__PURE__ */ i(
        "path",
        {
          fill: "url(#95yY7w43Oj6n2vH63j6HJe)",
          d: `M59,31.97c0.01,7.73-3.26,14.58-8.55,19.31l-7.62-6.72c2.1-1.61,3.77-3.71,4.84-6.15
        c0.29-0.66-0.2-1.41-0.92-1.41H37c-2.21,0-4-1.79-4-4v-2c0-2.21,1.79-4,4-4h17C56.75,27,59,29.22,59,31.97z`
        }
      )
    ]
  }
) }), Pr = (e) => /* @__PURE__ */ i(
  "svg",
  {
    width: 32,
    height: 32,
    viewBox: "0 0 56 56",
    style: { transform: "scale(2.8)" },
    version: "1.1",
    xmlns: "http://www.w3.org/2000/svg",
    children: /* @__PURE__ */ i(
      "g",
      {
        stroke: e === "light" ? "#424245" : "white",
        strokeWidth: "0.5",
        fillRule: "evenodd",
        children: /* @__PURE__ */ i(
          "path",
          {
            d: "M28.2226562,20.3846154 C29.0546875,20.3846154 30.0976562,19.8048315 30.71875,19.0317864 C31.28125,18.3312142 31.6914062,17.352829 31.6914062,16.3744437 C31.6914062,16.2415766 31.6796875,16.1087095 31.65625,16 C30.7304687,16.0362365 29.6171875,16.640178 28.9492187,17.4494596 C28.421875,18.06548 27.9414062,19.0317864 27.9414062,20.0222505 C27.9414062,20.1671964 27.9648438,20.3121424 27.9765625,20.3604577 C28.0351562,20.3725366 28.1289062,20.3846154 28.2226562,20.3846154 Z M25.2929688,35 C26.4296875,35 26.9335938,34.214876 28.3515625,34.214876 C29.7929688,34.214876 30.109375,34.9758423 31.375,34.9758423 C32.6171875,34.9758423 33.4492188,33.792117 34.234375,32.6325493 C35.1132812,31.3038779 35.4765625,29.9993643 35.5,29.9389701 C35.4179688,29.9148125 33.0390625,28.9122695 33.0390625,26.0979021 C33.0390625,23.6579784 34.9140625,22.5588048 35.0195312,22.474253 C33.7773438,20.6382708 31.890625,20.5899555 31.375,20.5899555 C29.9804688,20.5899555 28.84375,21.4596313 28.1289062,21.4596313 C27.3554688,21.4596313 26.3359375,20.6382708 25.1289062,20.6382708 C22.8320312,20.6382708 20.5,22.5950413 20.5,26.2911634 C20.5,28.5861411 21.3671875,31.013986 22.4335938,32.5842339 C23.3476562,33.9129053 24.1445312,35 25.2929688,35 Z",
            fill: e === "light" ? "#424245" : "white",
            fillRule: "nonzero"
          }
        )
      }
    )
  }
), Sr = (e) => /* @__PURE__ */ i(
  "svg",
  {
    fill: e === "light" ? "#1c1e21" : "white",
    role: "img",
    viewBox: "0 0 24 24",
    width: 28,
    height: 28,
    xmlns: "http://www.w3.org/2000/svg",
    children: /* @__PURE__ */ i(
      "path",
      {
        d: "M12 .297c-6.63 0-12 5.373-12 12 0 5.303 3.438 9.8 8.205 11.385.6.113.82-.258.82-.577 0-.285-.01-1.04-.015-2.04-3.338.724-4.042-1.61-4.042-1.61C4.422 18.07 3.633 17.7 3.633 17.7c-1.087-.744.084-.729.084-.729 1.205.084 1.838 1.236 1.838 1.236 1.07 1.835 2.809 1.305 3.495.998.108-.776.417-1.305.76-1.605-2.665-.3-5.466-1.332-5.466-5.93 0-1.31.465-2.38 1.235-3.22-.135-.303-.54-1.523.105-3.176 0 0 1.005-.322 3.3 1.23.96-.267 1.98-.399 3-.405 1.02.006 2.04.138 3 .405 2.28-1.552 3.285-1.23 3.285-1.23.645 1.653.24 2.873.12 3.176.765.84 1.23 1.91 1.23 3.22 0 4.61-2.805 5.625-5.475 5.92.42.36.81 1.096.81 2.22 0 1.606-.015 2.896-.015 3.286 0 .315.21.69.825.57C20.565 22.092 24 17.592 24 12.297c0-6.627-5.373-12-12-12"
      }
    )
  }
), xr = (e) => /* @__PURE__ */ i(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    width: 28,
    height: 28,
    viewBox: "0 0 90 90",
    children: /* @__PURE__ */ i("g", { children: /* @__PURE__ */ i(
      "path",
      {
        d: "M90,15.001C90,7.119,82.884,0,75,0H15C7.116,0,0,7.119,0,15.001v59.998   C0,82.881,7.116,90,15.001,90H45V56H34V41h11v-5.844C45,25.077,52.568,16,61.875,16H74v15H61.875C60.548,31,59,32.611,59,35.024V41   h15v15H59v34h16c7.884,0,15-7.119,15-15.001V15.001z",
        fill: e === "light" ? "#39569c" : "white"
      }
    ) })
  }
), kr = (e) => /* @__PURE__ */ i(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    width: 28,
    height: 28,
    viewBox: "0 0 480 480",
    children: /* @__PURE__ */ i("g", { children: /* @__PURE__ */ i(
      "path",
      {
        d: "M0.176,224L0.001,67.963l192-26.072V224H0.176z M224.001,37.241L479.937,0v224H224.001V37.241z M479.999,256l-0.062,224   l-255.936-36.008V256H479.999z M192.001,439.918L0.157,413.621L0.147,256h191.854V439.918z",
        fill: e === "light" ? "#00a2ed" : "white"
      }
    ) })
  }
), Lr = (e) => /* @__PURE__ */ i(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    width: 28,
    height: 28,
    viewBox: "0 0 24 24",
    children: /* @__PURE__ */ i(
      "path",
      {
        fill: e === "light" ? "#00acee" : "white",
        d: "M24 4.557c-.883.392-1.832.656-2.828.775 1.017-.609 1.798-1.574 2.165-2.724-.951.564-2.005.974-3.127 1.195-.897-.957-2.178-1.555-3.594-1.555-3.179 0-5.515 2.966-4.797 6.045-4.091-.205-7.719-2.165-10.148-5.144-1.29 2.213-.669 5.108 1.523 6.574-.806-.026-1.566-.247-2.229-.616-.054 2.281 1.581 4.415 3.949 4.89-.693.188-1.452.232-2.224.084.626 1.956 2.444 3.379 4.6 3.419-2.07 1.623-4.678 2.348-7.29 2.04 2.179 1.397 4.768 2.212 7.548 2.212 9.142 0 14.307-7.721 13.995-14.646.962-.695 1.797-1.562 2.457-2.549z"
      }
    )
  }
);
function Dr({
  allowSkipLogin: e,
  logo: t,
  signInOptions: n,
  firebaseApp: r,
  authController: s,
  noUserComponent: l,
  disableSignupScreen: c = !1,
  disabled: d = !1,
  additionalComponent: v,
  notAllowedError: a
}) {
  const o = Nt(), [u, g] = L(!1), [h, f] = L(!1), w = n.map((y) => typeof y == "object" ? y.provider : y), b = E(() => {
    const y = te(r), x = new Ye(y, "recaptcha", { size: "invisible" }), I = vt(y, s.authProviderError);
    if (I.hints[0].factorId === Te.FACTOR_ID) {
      const j = {
        multiFactorHint: I.hints[0],
        session: I.session
      };
      new Oe(y).verifyPhoneNumber(j, x).then(function(z) {
        const k = String(window.prompt("Please enter the verification code that was sent to your mobile device.")), K = Oe.credential(z, k), Y = Te.assertion(K);
        return I.resolveSignIn(Y);
      });
    } else
      console.warn("Unsupported second factor.");
  }, [s.authProviderError]);
  function P() {
    let y;
    if (s.user != null)
      return y;
    const x = ["auth/popup-closed-by-user", "auth/cancelled-popup-request"];
    return s.authProviderError && (s.authProviderError.code === "auth/operation-not-allowed" || s.authProviderError.code === "auth/configuration-not-found" ? y = /* @__PURE__ */ D(oe, { children: [
      /* @__PURE__ */ i("div", { className: "p-4", children: /* @__PURE__ */ i(
        Z,
        {
          title: "Firebase Auth not enabled",
          error: "You need to enable Firebase Auth and the corresponding login provider in your Firebase project"
        }
      ) }),
      r && /* @__PURE__ */ i("div", { className: "p-4", children: /* @__PURE__ */ i(
        "a",
        {
          href: `https://console.firebase.google.com/project/${r.options.projectId}/authentication/providers`,
          rel: "noopener noreferrer",
          target: "_blank",
          children: /* @__PURE__ */ i(
            A,
            {
              variant: "text",
              color: "error",
              children: "Open Firebase configuration"
            }
          )
        }
      ) })
    ] }) : s.authProviderError.code === "auth/invalid-api-key" ? y = /* @__PURE__ */ i("div", { className: "p-4", children: /* @__PURE__ */ i(
      Z,
      {
        title: "Invalid API key",
        error: "auth/invalid-api-key: Check that your Firebase config is set correctly in your `firebase_config.ts` file"
      }
    ) }) : x.includes(s.authProviderError.code) || (s.authProviderError.code === "auth/multi-factor-auth-required" && b(), y = /* @__PURE__ */ i("div", { className: "p-4", children: /* @__PURE__ */ i(Z, { error: s.authProviderError }) }))), y;
  }
  let C;
  t ? C = /* @__PURE__ */ i(
    "img",
    {
      src: t,
      style: {
        height: "100%",
        width: "100%",
        objectFit: "contain"
      },
      alt: "Logo"
    }
  ) : C = /* @__PURE__ */ i(It, {});
  let F;
  return a && (typeof a == "string" ? F = a : a instanceof Error ? F = a.message : F = "It looks like you don't have access to the CMS, based on the specified Authenticator configuration"), /* @__PURE__ */ D(
    "div",
    {
      className: "flex flex-col items-center justify-center min-h-screen min-w-full p-4",
      children: [
        /* @__PURE__ */ i("div", { id: "recaptcha" }),
        /* @__PURE__ */ D(
          "div",
          {
            className: "flex flex-col items-center w-full max-w-[500px]",
            children: [
              /* @__PURE__ */ i("div", { className: "p-1 w-64 h-64 m-4", children: C }),
              F && /* @__PURE__ */ i("div", { className: "p-8", children: /* @__PURE__ */ i(Z, { error: F }) }),
              P(),
              !u && !h && /* @__PURE__ */ D("div", { className: "my-4 w-full", children: [
                nt(s, w, o.mode, d),
                w.includes("password") && /* @__PURE__ */ i(
                  B,
                  {
                    disabled: d,
                    text: "Email/password",
                    icon: /* @__PURE__ */ i(rr, { size: 28 }),
                    onClick: () => g(!0)
                  }
                ),
                w.includes("phone") && /* @__PURE__ */ i(
                  B,
                  {
                    disabled: d,
                    text: "Phone number",
                    icon: /* @__PURE__ */ i(or, { size: 28 }),
                    onClick: () => f(!0)
                  }
                ),
                w.includes("anonymous") && /* @__PURE__ */ i(
                  B,
                  {
                    disabled: d,
                    text: "Log in anonymously",
                    icon: /* @__PURE__ */ i(
                      nr,
                      {
                        size: 28
                      }
                    ),
                    onClick: s.anonymousLogin
                  }
                ),
                e && /* @__PURE__ */ i(
                  A,
                  {
                    className: "m-1 mb-4",
                    variant: "text",
                    disabled: d,
                    onClick: s.skipLogin,
                    children: "Skip login"
                  }
                )
              ] }),
              u && /* @__PURE__ */ i(
                Nr,
                {
                  authController: s,
                  onClose: () => g(!1),
                  mode: o.mode,
                  noUserComponent: l,
                  disableSignupScreen: c
                }
              ),
              h && /* @__PURE__ */ i(
                jr,
                {
                  authController: s,
                  onClose: () => f(!1)
                }
              ),
              !u && !h && v
            ]
          }
        )
      ]
    }
  );
}
function B({
  icon: e,
  onClick: t,
  text: n,
  disabled: r
}) {
  return /* @__PURE__ */ i("div", { className: "m-1 w-full", children: /* @__PURE__ */ i(
    A,
    {
      className: "w-full",
      variant: "outlined",
      disabled: r,
      onClick: t,
      children: /* @__PURE__ */ D(
        "div",
        {
          className: "p-1 flex h-8 items-center justify-items-center",
          children: [
            /* @__PURE__ */ i(
              "div",
              {
                className: "flex flex-col w-8 items-center justify-items-center",
                children: e
              }
            ),
            /* @__PURE__ */ i("div", { className: "flex-grow pl-2 text-center", children: n })
          ]
        }
      )
    }
  ) });
}
function jr({
  onClose: e,
  authController: t
}) {
  Cr();
  const [n, r] = L(), [s, l] = L(), [c, d] = L(!1);
  return /* @__PURE__ */ D("form", { onSubmit: async (a) => {
    a.preventDefault(), s && t.confirmationResult ? (d(!1), t.confirmationResult.confirm(s).catch((o) => {
      o.code === "auth/invalid-verification-code" && d(!0);
    })) : n && t.phoneLogin(n, window.recaptchaVerifier);
  }, children: [
    c && /* @__PURE__ */ i("div", { className: "p-8", children: /* @__PURE__ */ i(Z, { error: "Invalid confirmation code" }) }),
    /* @__PURE__ */ i("div", { id: ot }),
    /* @__PURE__ */ D("div", { className: "flex flex-col gap-1", children: [
      /* @__PURE__ */ i(
        Ze,
        {
          onClick: e,
          children: /* @__PURE__ */ i(Xe, { className: "w-5 h-5" })
        }
      ),
      /* @__PURE__ */ i("div", { className: "p-1 flex", children: /* @__PURE__ */ i(
        X,
        {
          align: "center",
          variant: "subtitle2",
          children: "Please enter your phone number"
        }
      ) }),
      /* @__PURE__ */ i(
        le,
        {
          placeholder: "",
          value: n ?? "",
          disabled: !!(n && (t.authLoading || t.confirmationResult)),
          type: "phone",
          onChange: (a) => r(a.target.value)
        }
      ),
      !!(n && t.confirmationResult) && /* @__PURE__ */ D(oe, { children: [
        /* @__PURE__ */ i("div", { className: "mt-2 p-1 flex", children: /* @__PURE__ */ i(
          X,
          {
            align: "center",
            variant: "subtitle2",
            children: "Please enter the confirmation code"
          }
        ) }),
        /* @__PURE__ */ i(
          le,
          {
            placeholder: "",
            value: s ?? "",
            type: "text",
            onChange: (a) => l(a.target.value)
          }
        )
      ] }),
      /* @__PURE__ */ D("div", { className: "flex justify-end items-center w-full", children: [
        t.authLoading && /* @__PURE__ */ i(et, { className: "p-1", size: "small" }),
        /* @__PURE__ */ i(A, { type: "submit", children: "Ok" })
      ] })
    ] })
  ] });
}
function Nr({
  onClose: e,
  authController: t,
  mode: n,
  noUserComponent: r,
  disableSignupScreen: s
}) {
  const l = xe(null), [c, d] = L("email"), [v, a] = L(), [o, u] = L(), [g, h] = L();
  U(() => {
    (c === "password" || c === "registration") && l.current && l.current.focus();
  }, [c]), U(() => {
    if (!document)
      return;
    const y = (x) => {
      x.keyCode === 27 && e();
    };
    return document.addEventListener("keydown", y, !1), () => {
      document.removeEventListener("keydown", y, !1);
    };
  }, [e]);
  function f() {
    v && (t.fetchSignInMethodsForEmail(v).then((y) => {
      h(y.filter((x) => x !== "password"));
    }), d("password"));
  }
  function w() {
    v && o && t.emailPasswordLogin(v, o);
  }
  function b() {
    v && o && t.createUserWithEmailAndPassword(v, o);
  }
  return /* @__PURE__ */ i(
    "form",
    {
      className: "w-full",
      onSubmit: (y) => {
        y.preventDefault(), c === "email" ? f() : c === "password" ? w() : c === "registration" && b();
      },
      children: /* @__PURE__ */ D("div", { className: "max-w-[480px] w-full flex flex-col gap-4", children: [
        /* @__PURE__ */ i(
          Ze,
          {
            onClick: () => {
              c === "email" ? e() : c === "password" || c === "registration" ? d("email") : h(void 0);
            },
            children: /* @__PURE__ */ i(Xe, { className: "w-5 h-5" })
          }
        ),
        /* @__PURE__ */ i("div", { children: c === "registration" && r }),
        /* @__PURE__ */ i(
          X,
          {
            className: `${c === "registration" && s ? "hidden" : "flex"}`,
            variant: "subtitle2",
            children: c === "registration" ? "Please enter your email and password to create an account" : c === "password" ? "Please enter your password" : "Please enter your email"
          }
        ),
        (c === "email" || c === "registration") && /* @__PURE__ */ i(
          le,
          {
            placeholder: "Email",
            autoFocus: !0,
            value: v ?? "",
            disabled: t.authLoading,
            type: "email",
            onChange: (y) => a(y.target.value)
          }
        ),
        /* @__PURE__ */ i(
          "div",
          {
            className: `${c === "password" || c === "registration" && !s ? "block" : "hidden"}`,
            children: /* @__PURE__ */ i(
              le,
              {
                placeholder: "Password",
                value: o ?? "",
                disabled: t.authLoading,
                inputRef: l,
                type: "password",
                onChange: (y) => u(y.target.value)
              }
            )
          }
        ),
        /* @__PURE__ */ D(
          "div",
          {
            className: `${c === "registration" && s ? "hidden" : "flex"} justify-end items-center w-full flex gap-2`,
            children: [
              t.authLoading && /* @__PURE__ */ i(et, { className: "p-1", size: "small" }),
              c === "email" && /* @__PURE__ */ i(A, { variant: "outlined", onClick: () => d("registration"), children: "New user" }),
              /* @__PURE__ */ i(A, { type: "submit", children: c === "registration" ? "Create account" : "Login" })
            ]
          }
        ),
        g && g.length > 0 && /* @__PURE__ */ D("div", { className: "flex flex-col gap-4 p-4", children: [
          /* @__PURE__ */ D("div", { children: [
            /* @__PURE__ */ i(X, { variant: "subtitle2", children: "You already have an account" }),
            /* @__PURE__ */ D(X, { variant: "body2", children: [
              "You can use one of these methods to login with ",
              v
            ] })
          ] }),
          /* @__PURE__ */ i("div", { children: g && nt(t, g, n, !1) })
        ] })
      ] })
    }
  );
}
function nt(e, t, n, r) {
  return /* @__PURE__ */ D(oe, { children: [
    t.includes("google.com") && /* @__PURE__ */ i(
      B,
      {
        disabled: r,
        text: "Sign in with Google",
        icon: Fr(),
        onClick: e.googleLogin
      }
    ),
    t.includes("microsoft.com") && /* @__PURE__ */ i(
      B,
      {
        disabled: r,
        text: "Sign in with Microsoft",
        icon: kr(n),
        onClick: e.microsoftLogin
      }
    ),
    t.includes("apple.com") && /* @__PURE__ */ i(
      B,
      {
        disabled: r,
        text: "Sign in with Apple",
        icon: Pr(n),
        onClick: e.appleLogin
      }
    ),
    t.includes("github.com") && /* @__PURE__ */ i(
      B,
      {
        disabled: r,
        text: "Sign in with Github",
        icon: Sr(n),
        onClick: e.githubLogin
      }
    ),
    t.includes("facebook.com") && /* @__PURE__ */ i(
      B,
      {
        disabled: r,
        text: "Sign in with Facebook",
        icon: xr(n),
        onClick: e.facebookLogin
      }
    ),
    t.includes("twitter.com") && /* @__PURE__ */ i(
      B,
      {
        disabled: r,
        text: "Sign in with Twitter",
        icon: Lr(n),
        onClick: e.twitterLogin
      }
    )
  ] });
}
const qe = [
  Ge.PROVIDER_ID
];
function to({
  name: e,
  logo: t,
  logoDark: n,
  toolbarExtraWidget: r,
  authenticator: s,
  collections: l,
  views: c,
  adminViews: d,
  textSearchControllerBuilder: v,
  allowSkipLogin: a,
  signInOptions: o = qe,
  firebaseConfig: u,
  onFirebaseInit: g,
  appCheckOptions: h,
  dateTimeFormat: f,
  locale: w,
  basePath: b,
  baseCollectionPath: P,
  onAnalyticsEvent: C,
  propertyConfigs: F,
  plugins: y,
  autoOpenDrawer: x,
  firestoreIndexesBuilder: I,
  components: j,
  localTextSearchEnabled: H = !1
}) {
  Tt(e, t);
  const z = (F ?? []).map((R) => ({
    [R.key]: R
  })).reduce((R, se) => ({ ...R, ...se }), {}), {
    firebaseApp: k,
    firebaseConfigLoading: K,
    configError: Y
  } = ur({
    onFirebaseInit: g,
    firebaseConfig: u
  }), ne = Ot(), {
    loading: de,
    appCheckVerified: De,
    error: je
  } = dr({
    firebaseApp: k,
    options: h
  }), N = cr({
    firebaseApp: k,
    signInOptions: o
  }), fe = Rt(), Q = wr({
    firebaseApp: k,
    textSearchControllerBuilder: v,
    firestoreIndexesBuilder: I,
    localTextSearchEnabled: H
  }), ie = ar({
    firebaseApp: k
  }), {
    authLoading: he,
    canAccessMainView: ge,
    notAllowedError: me
  } = Mt({
    authController: N,
    authenticator: s,
    dataSourceDelegate: Q,
    storageSource: ie
  }), pe = _t({
    collections: l,
    views: c,
    basePath: b,
    baseCollectionPath: P,
    authController: N,
    adminViews: d,
    dataSourceDelegate: Q
  });
  return K || !k || de ? /* @__PURE__ */ i(oe, { children: /* @__PURE__ */ i(Me, {}) }) : Y ? /* @__PURE__ */ i(ir, { children: Y }) : /* @__PURE__ */ i(Ut, { children: /* @__PURE__ */ i(Ht, { value: ne, children: /* @__PURE__ */ i(
    Vt,
    {
      authController: N,
      navigationController: pe,
      userConfigPersistence: fe,
      dateTimeFormat: f,
      dataSourceDelegate: Q,
      storageSource: ie,
      entityLinkBuilder: ({ entity: R }) => `https://console.firebase.google.com/project/${k.options.projectId}/firestore/data/${R.path}/${R.id}`,
      locale: w,
      onAnalyticsEvent: C,
      plugins: y,
      propertyConfigs: z,
      children: ({
        context: R,
        loading: se
      }) => {
        let m;
        if (se || he)
          m = /* @__PURE__ */ i(Me, { size: "large" });
        else {
          const p = ne.mode === "dark" && n ? n : t;
          if (ge)
            m = /* @__PURE__ */ D(
              Bt,
              {
                name: e,
                logo: p,
                fireCMSAppBarProps: {
                  endAdornment: r
                },
                autoOpenDrawer: x,
                children: [
                  /* @__PURE__ */ i(
                    $t,
                    {
                      HomePage: j?.HomePage
                    }
                  ),
                  /* @__PURE__ */ i(zt, {})
                ]
              }
            );
          else {
            const S = j?.LoginView ?? Dr;
            m = /* @__PURE__ */ i(
              S,
              {
                logo: p,
                allowSkipLogin: a,
                signInOptions: o ?? qe,
                firebaseApp: k,
                authController: N,
                notAllowedError: me
              }
            );
          }
        }
        return m;
      }
    }
  ) }) });
}
async function it(e, t, n = !1, r, s = 100) {
  if (r && (r ?? [])?.length > 0) {
    const [d, ...v] = r, a = await G($(_(e, d), W(5)));
    return Promise.all(a.docs.map((o) => it(
      e,
      o.ref.path + "/" + t,
      n,
      v,
      Math.max(Math.ceil(s / 5), 5)
    ))).then((o) => o.flat());
  }
  const l = $(
    n ? We(e, t) : _(e, t),
    W(s)
  ), c = await G(l);
  return console.debug("Got documents", n, l, c.docs), c.docs;
}
async function Ir(e, t, n, r) {
  console.debug("Building schema for collection", t, r);
  const s = Je(t), c = (await it(e, s, n, r)).map((d) => d.data()).filter(Boolean);
  return Tr(t, c);
}
async function Tr(e, t) {
  const n = Je(e), r = await tt(t, st), s = sr(r), l = n.includes("/") ? n.split("/").slice(-1)[0] : n;
  return {
    path: n,
    name: qt(l),
    properties: r,
    propertiesOrder: s
  };
}
async function ro(e) {
  return tt(e, st);
}
function st(e) {
  return typeof e == "number" ? "number" : typeof e == "string" ? "string" : typeof e == "boolean" ? "boolean" : Array.isArray(e) ? "array" : e instanceof re ? "date" : e instanceof ke ? "geopoint" : e instanceof Qe ? "reference" : "map";
}
const oo = (e) => {
  if (e)
    return (t, n, r) => {
      const s = O(e);
      return Ir(s, t, n, r);
    };
};
export {
  to as FireCMSFirebaseApp,
  Dr as FirebaseLoginView,
  B as LoginButton,
  ot as RECAPTCHA_CONTAINER_ID,
  Qr as buildAlgoliaSearchController,
  Yr as buildCollectionId,
  oo as buildCollectionInference,
  Xr as buildPineconeSearchController,
  ee as cmsToFirestoreModel,
  Se as cmsToRTDBModel,
  fr as docToCollection,
  Jr as docsToCollectionTree,
  ue as firestoreToCMSModel,
  Ar as getFirestoreDataInPath,
  Ir as getInferredEntityCollection,
  Tr as getInferredEntityCollectionFromData,
  ro as getPropertiesFromData,
  ze as localSearchControllerBuilder,
  Kr as performAlgoliaTextSearch,
  Zr as performPineconeTextSearch,
  Wr as prepareCollectionForPersistence,
  Fe as setUndefinedToDelete,
  dr as useAppCheck,
  cr as useFirebaseAuthController,
  eo as useFirebaseRTDBDelegate,
  ar as useFirebaseStorageSource,
  wr as useFirestoreDelegate,
  ur as useInitialiseFirebase,
  Cr as useRecaptcha
};
