// @firecms/cloud/dist/index.es.js
import z, { useState as x, useRef as ve, useEffect as k, useCallback as j, useContext as xt, useDeferredValue as Ct, useMemo as Pe } from "react";
import { getFirestore as G, onSnapshot as Q, doc as J, setDoc as se, collection as ae, query as Ce, where as we, getDocs as wt, getDoc as Ne, addDoc as bt, deleteDoc as St } from "firebase/firestore";
import { getStorage as Pt, getDownloadURL as Nt, uploadBytes as Et, ref as _t } from "firebase/storage";
import { getUserRoles as Lt, resolveUserRolePermissions as Ge, UserManagementProvider as Tt, UsersView as Ft, RolesView as Ut } from "@firecms/user_management";
import { jsx as e, jsxs as i, Fragment as A } from "react/jsx-runtime";
import { getAuth as pe, signInWithCustomToken as jt, onAuthStateChanged as kt, GoogleAuthProvider as Me, signInWithPopup as Rt, signOut as At, OAuthCredential as Dt } from "firebase/auth";
import { useNavigationController as he, useAuthController as We, DrawerNavigationItem as Ot, IconForView as Ee, useSnackbarController as Ue, useBrowserTitleAndIcon as je, FireCMSLogo as qe, ErrorView as le, useCustomizationController as Mt, DefaultHomePage as It, NavigationGroup as zt, SmallNavigationCard as $t, unslugify as Vt, useBuildModeController as Bt, Scaffold as ye, useBuildLocalConfigurationPersistence as Ht, useBuildNavigationController as Jt, SnackbarProvider as Yt, ModeControllerProvider as Gt, FireCMS as Wt, CircularProgressCenter as q, NavigationRoutes as qt, SideDialogs as Qt } from "@firecms/core";
export * from "@firecms/core";
import { Typography as g, Tooltip as Ie, Button as V, AddIcon as Qe, Menu as Kt, IconButton as ke, MoreVertIcon as Xt, cn as re, MenuItem as Zt, Paper as Re, LoadingButton as fe, BuildIcon as Ae, Chip as K, Select as er, SelectItem as tr, RocketLaunchIcon as Ke, Alert as De, Dialog as Xe, DialogContent as Ze, CloseIcon as et, Container as rr, CircularProgress as tt, Card as or, AutoAwesomeIcon as Z, defaultBorderMixin as nr, CenteredView as de, BooleanSwitchWithLabel as _e, TextField as ir, FileUpload as sr, BooleanSwitch as ar, Checkbox as ze, SearchIcon as lr, Collapse as cr, ErrorIcon as rt } from "@firecms/ui";
export * from "@firecms/ui";
import { useCollectionEditorController as ot, mergeCollections as dr, useCollectionEditorPlugin as ur, MissingReferenceWidget as fr } from "@firecms/collection_editor";
import { useNavigate as nt, BrowserRouter as hr, Route as mr } from "react-router-dom";
import { useDataEnhancementPlugin as gr } from "@firecms/data_enhancement";
import { useInitialiseFirebase as it, useAppCheck as pr, useFirebaseAuthController as yr, useFirestoreDelegate as vr, useFirebaseStorageSource as xr, buildCollectionInference as Cr, getFirestoreDataInPath as wr } from "@firecms/firebase";
export * from "@firecms/firebase";
import { useImportExportPlugin as br } from "@firecms/data_import_export";
import { useFirestoreCollectionsConfigController as Sr } from "@firecms/collection_editor_firebase";
const st = ["Admin"];
function Pr({
  user: t,
  userManagement: r,
  collection: o
}) {
  const n = {
    createCollections: !1,
    editCollections: !1,
    deleteCollections: !1
  }, a = t && r.users.find((s) => s.uid === t.uid), l = a ? Lt(r.roles, a) : void 0;
  return !a || !l ? n : l.map((s) => ({
    createCollections: s.isAdmin || s.config?.createCollections === !0,
    editCollections: s.isAdmin || s.config?.editCollections === !0 || s.config?.editCollections === "own" && o?.ownerId === a?.uid,
    deleteCollections: s.isAdmin || s.config?.deleteCollections === !0 || s.config?.deleteCollections === "own" && o?.ownerId === a?.uid
  })).reduce(Nr, n);
}
const Nr = (t, r) => ({
  createCollections: t.createCollections || r.createCollections,
  editCollections: t.editCollections || r.editCollections,
  deleteCollections: t.deleteCollections || r.deleteCollections
}), zo = (t, r) => t.length === r.length && t.every((o) => r.includes(o)), at = [
  {
    path: "users",
    name: "CMS Users",
    group: "Admin",
    icon: "face"
  },
  {
    path: "roles",
    name: "Roles",
    group: "Admin",
    icon: "gpp_good"
  },
  {
    path: "settings",
    name: "Project settings",
    group: "Admin",
    icon: "settings"
  }
];
function Er(t, r) {
  if (!r)
    return;
  const o = Tr(r), n = new Date(o.exp * 1e3);
  localStorage.setItem(`auth_token::${t}`, JSON.stringify({
    token: r,
    expiry: n
  }));
}
function _r(t) {
  const r = localStorage.getItem(`auth_token::${t}`);
  if (r) {
    const o = JSON.parse(r);
    if (o.expiry = new Date(o.expiry), o.expiry > /* @__PURE__ */ new Date())
      return o.token;
  }
}
function Lr() {
  for (let t = 0; t < localStorage.length; t++) {
    const r = localStorage.key(t);
    r?.startsWith("auth_token::") && localStorage.removeItem(r);
  }
}
function Tr(t) {
  if (!t)
    throw new Error("No JWT token");
  const o = t.split(".")[1].replace(/-/g, "+").replace(/_/g, "/"), n = decodeURIComponent(window.atob(o).split("").map(function(a) {
    return "%" + ("00" + a.charCodeAt(0).toString(16)).slice(-2);
  }).join(""));
  return JSON.parse(n);
}
function be(t, r = 10) {
  if (!/^#([0-9A-Fa-f]{3}){1,2}$/.test(t))
    throw new Error("Invalid color format");
  let o = t.substring(1).split("");
  o.length === 3 && (o = [o[0], o[0], o[1], o[1], o[2], o[2]]);
  let n = parseInt(o[0] + o[1], 16), a = parseInt(o[2] + o[3], 16), l = parseInt(o[4] + o[5], 16);
  return n = Math.floor(n * (1 - r / 100)), a = Math.floor(a * (1 - r / 100)), l = Math.floor(l * (1 - r / 100)), "#" + (n < 16 ? "0" : "") + n.toString(16) + (a < 16 ? "0" : "") + a.toString(16) + (l < 16 ? "0" : "") + l.toString(16);
}
function $e(t, r = 10) {
  if (!/^#([0-9A-Fa-f]{3}){1,2}$/.test(t))
    throw new Error("Invalid color format");
  let o = t.substring(1).split("");
  o.length === 3 && (o = [o[0], o[0], o[1], o[1], o[2], o[2]]);
  const n = parseInt(o[0] + o[1], 16), a = parseInt(o[2] + o[3], 16), l = parseInt(o[4] + o[5], 16), s = r / 100;
  return `rgba(${n}, ${a}, ${l}, ${s})`;
}
const ie = "#0070F4", ge = "#FF5B79";
function Fr({
  backendFirebaseApp: t,
  projectId: r
}) {
  const [o, n] = x(ie), [a, l] = x(ge), s = r ? `projects/${r}` : void 0, [d, p] = x(), [u, h] = x(), [S, v] = x(!1), [c, m] = x(), [y, f] = x(), [E, T] = x(), [C, P] = x(), [F, O] = x(!1), [U, D] = x(), [$, B] = x(), N = ve(r), [R, b] = z.useState();
  k(() => {
    if (!t)
      throw Error("useBuildProjectConfig Firebase not initialised");
    const I = G(t);
    if (!(!I || !s))
      return Q(
        J(I, s),
        {
          next: (L) => {
            b(L.get("logo"));
          },
          error: (L) => {
            console.error(L);
          }
        }
      );
  }, [s]), k(() => {
    o ? (document.documentElement.style.setProperty("--fcms-primary", o), document.documentElement.style.setProperty("--fcms-primary-dark", be(o, 10)), document.documentElement.style.setProperty("--fcms-primary-bg", $e(o, 10))) : (document.documentElement.style.setProperty("--fcms-primary", be(ie, 10)), document.documentElement.style.setProperty("--fcms-primary-dark", be(ie, 10)), document.documentElement.style.setProperty("--fcms-primary-bg", $e(ie, 10))), a ? document.documentElement.style.setProperty("--fcms-secondary", a) : document.documentElement.style.setProperty("--fcms-secondary", ge);
  }, [o, a]);
  const w = j(async (I) => {
    if (!t)
      throw Error("useBuildProjectConfig Firebase not initialised");
    const L = G(t);
    if (!L || !s)
      throw Error("useFirestoreConfigurationPersistence Firestore not initialised");
    const ue = Pt(t);
    if (!ue)
      throw Error("useFirestoreConfigurationPersistence Storage not initialised");
    const xe = await Ur(ue, {
      file: I,
      path: `${s}/images`
    }), ne = await Nt(xe);
    se(J(L, s), { logo: ne }, { merge: !0 });
  }, [s]), _ = j(async (I) => {
    if (!t)
      throw Error("useBuildProjectConfig Firebase not initialised");
    const L = G(t);
    if (!L || !s)
      throw Error("useFirestoreConfigurationPersistence Firestore not initialised");
    return se(J(L, s), { name: I }, { merge: !0 });
  }, [s]), M = j(async (I) => {
    if (!t)
      throw Error("useBuildProjectConfig Firebase not initialised");
    const L = G(t);
    if (!L || !s)
      throw Error("useFirestoreConfigurationPersistence Firestore not initialised");
    return se(J(L, s), { local_text_search_enabled: I }, { merge: !0 });
  }, [s]);
  k(() => {
    if (!r || !t) {
      v(!1);
      return;
    }
    N.current !== r && (v(!0), m(void 0), N.current = void 0);
    const I = G(t);
    return Q(
      J(I, "projects", r),
      {
        next: (L) => {
          console.log("Project config snapshot", L.data()), p(L.get("name"));
          const ue = L.get("subscription_plan") ?? "free";
          h(ue), O(L.get("local_text_search_enabled") ?? !1), ue === "free" ? (n(ie), l(ge)) : (n(L.get("primary_color") ?? ie), l(L.get("secondary_color") ?? ge));
          const xe = L.get("current_app_config_revision");
          D(xe), B(L.get("creation_type"));
          const ne = L.get("firebase_config");
          N.current = r, ne === "loading" ? (v(!0), m(void 0), f(!1)) : typeof ne == "object" ? (m(ne), v(!1), f(!1)) : ne === void 0 && (v(!1), f(!0)), P(void 0), T(!L.get("service_account"));
        },
        error: (L) => {
          console.error(L), P(L), v(!1);
        }
      }
    );
  }, [t, r]);
  const H = u === "free" ? 3 : void 0, Oe = u !== "free", W = u !== "free", me = u !== "free", pt = u !== "free", yt = j(async (I) => {
    if (!t)
      throw Error("useBuildProjectConfig Firebase not initialised");
    const L = G(t);
    if (!L || !s)
      throw Error("useFirestoreConfigurationPersistence Firestore not initialised");
    n(I), W && se(J(L, s), { primary_color: I }, { merge: !0 });
  }, [s, W]), vt = j(async (I) => {
    if (!t)
      throw Error("useBuildProjectConfig Firebase not initialised");
    const L = G(t);
    if (!L || !s)
      throw Error("useFirestoreConfigurationPersistence Firestore not initialised");
    l(I), W && se(J(L, s), { secondary_color: I }, { merge: !0 });
  }, [s, W]);
  return {
    projectId: r,
    logo: W ? R : void 0,
    uploadLogo: w,
    updateProjectName: _,
    projectName: d,
    subscriptionPlan: N.current !== r ? void 0 : u,
    customizationRevision: N.current !== r ? void 0 : U,
    configLoading: N.current !== r || S,
    configError: N.current !== r ? void 0 : C,
    clientFirebaseConfig: N.current !== r ? void 0 : c,
    clientFirebaseMissing: N.current !== r ? void 0 : y,
    serviceAccountMissing: N.current !== r ? void 0 : E,
    usersLimit: H,
    canEditRoles: Oe,
    canModifyTheme: W,
    canExport: me,
    canUseLocalTextSearch: pt,
    localTextSearchEnabled: F,
    updateLocalTextSearchEnabled: M,
    primaryColor: o,
    secondaryColor: a,
    updatePrimaryColor: yt,
    updateSecondaryColor: vt
  };
}
const Ur = (t, {
  file: r,
  fileName: o,
  path: n,
  metadata: a
}) => {
  const l = o ?? r.name;
  return console.debug("Uploading file", l, r, n, a), Et(_t(t, `${n}/${l}`), r, a).then((s) => s.ref);
}, lt = z.createContext({}), X = () => xt(lt);
function jr({ children: t, config: r }) {
  return /* @__PURE__ */ e(lt.Provider, { value: r, children: t });
}
function kr({
  projectsApi: t,
  firebaseApp: r,
  projectId: o,
  onUserChanged: n
}) {
  const [a, l] = x(!1), [s, d] = x(!1), [p, u] = x(void 0), h = j(async (S = !1) => {
    if (console.debug("Checking delegated login", { skipCache: S, projectId: o }), r && o) {
      u(void 0), d(!0), l(!1);
      try {
        let v = !1, c = !S && _r(o);
        if (c)
          console.debug("Using cached token", o), v = !0;
        else
          try {
            console.debug("Delegating login", o), c = await t.doDelegatedLogin(o);
          } catch (y) {
            console.error("Error delegating login", JSON.stringify(y)), u(y);
          }
        if (!c)
          return;
        const m = pe(r);
        jt(m, c).then(async (y) => {
          console.log("Delegated user signed in", y), n?.(y.user), l(!0);
          try {
            Er(o, c);
          } catch (f) {
            console.error("Error caching token", f);
          }
        }).catch(async (y) => {
          if (v)
            return await h(!0);
          console.error("Error signing in with delegated token", y), l(!1), u(y);
        }).finally(() => d(!1));
      } catch (v) {
        l(!1), u(v), d(!1);
      }
    }
  }, [r, o]);
  return k(() => {
    h();
  }, [h]), {
    loginSuccessful: a,
    delegatedLoginLoading: s,
    delegatedLoginError: p
  };
}
const ct = z.createContext({});
function Rr({
  children: t,
  ...r
}) {
  return /* @__PURE__ */ e(ct.Provider, { value: r, children: t });
}
function oe() {
  return z.useContext(ct);
}
const Ar = "subscriptions", Ve = "products", Be = "customers", Dr = "checkout_sessions";
function dt({
  firebaseApp: t
}) {
  const { backendUid: r } = oe(), o = ve(), [n, a] = x(), [l, s] = x(!0), [d, p] = x(), [u, h] = x([]), [S, v] = x(!0), [c, m] = x();
  k(() => {
    t && (o.current = G(t));
  }, [t]), k(() => {
    const C = o.current;
    if (!C)
      return;
    const P = ae(C, Ve);
    return Q(Ce(P, we("active", "==", !0)), {
      next: async (F) => {
        const O = await Promise.all(F.docs.map(async (U) => {
          const D = ae(C, Ve, U.id, "prices");
          return wt(Ce(D, we("active", "==", !0))).then(async ($) => {
            const B = $.docs.map((N) => ({
              ...N.data(),
              id: N.id
            }));
            return B.sort((N, R) => (R.default ? 1 : 0) - (N.default ? 1 : 0)), {
              id: U.id,
              ...U.data(),
              prices: B
            };
          });
        }));
        p(void 0), a(O), s(!1);
      },
      error: (F) => {
        p(F);
      }
    });
  }, [o]);
  function y(C) {
    return Ne(C).then((P) => {
      if (P.exists())
        return {
          id: P.id,
          ...P.data() ?? {}
        };
    });
  }
  function f(C) {
    return Ne(C).then((P) => {
      if (P.exists())
        return {
          id: P.id,
          ...P.data() ?? {}
        };
    });
  }
  return k(() => {
    const C = o.current;
    if (!C || !r)
      return;
    const P = ae(C, Be, r, Ar);
    return Q(
      Ce(P, we("ended_at", "==", null)),
      {
        next: async (F) => {
          const O = (await Promise.all(F.docs.map(async (U) => {
            const [D, $] = await Promise.all([y(U.data().price), f(U.data().product)]);
            if (!D) {
              console.warn("Price not found for subscription", U.id);
              return;
            }
            return {
              id: U.id,
              ...U.data(),
              price: D,
              product: $
            };
          }))).filter(Boolean);
          v(!1), m(void 0), h(O);
        },
        error: (F) => {
          m(F);
        }
      }
    );
  }, [o, r]), {
    products: n,
    subscribe: async (C, P, F, O) => {
      const U = o.current;
      if (!U)
        throw new Error("Firestore not initialized");
      if (!r)
        throw new Error("User not logged in");
      if (!O)
        throw new Error("subscription type not provided. Make sure to assign the metadata.type field in the Stripe dashboard to a product.");
      const D = {
        price: P.id
      };
      P.recurring?.usage_type !== "metered" && (D.quantity = 1);
      const $ = {
        automatic_tax: !0,
        tax_id_collection: !0,
        collect_shipping_address: !1,
        allow_promotion_codes: !0,
        line_items: [D],
        success_url: `${window.location.origin}${window.location.pathname}`,
        cancel_url: `${window.location.origin}${window.location.pathname}`,
        metadata: {
          projectId: C,
          type: O
        }
      };
      P.type === "one_time" && ($.mode = "payment", $.payment_method_types = ["card", "sepa_debit", "sofort"]);
      const B = ae(U, Be, r, Dr), N = await bt(B, $);
      return Q(N, (R) => {
        const {
          error: b,
          url: w
        } = R.data();
        F(w, b);
      });
    },
    getSubscriptionsForProject: (C) => u?.filter((P) => P.metadata.projectId === C),
    activeSubscriptions: u,
    productsLoading: l,
    productsLoadingError: d,
    activeSubscriptionsLoading: S,
    activeSubscriptionsLoadingError: c
  };
}
class Le extends Error {
  code;
  projectId;
  data;
  constructor(r, o, n, a) {
    super(r), this.code = o, this.projectId = n, this.data = a;
  }
}
async function ee(t, r) {
  const o = await t.json();
  if (t.ok)
    return o.data;
  throw t.status === 409 ? new Le(o.message ?? o.error ?? "This already exists.", o.code ?? "already-exists", r) : new Le(o.message ?? o.error, o.code, r, o.data);
}
const Se = {};
function Or(t, r) {
  async function o(c, m, y, f) {
    const E = await r();
    return fetch(
      t + "/projects",
      {
        method: "POST",
        headers: Y({ firebaseAccessToken: E, googleAccessToken: m, serviceAccount: y }),
        body: JSON.stringify({
          projectId: c,
          creationType: f,
          serviceAccount: y ?? null
        })
      }
    ).then(async (T) => ee(T, c).then((C) => !0));
  }
  async function n(c) {
    const m = await r();
    return fetch(
      t + `/projects/${c}/firebase_webapp`,
      {
        method: "POST",
        headers: Y({ firebaseAccessToken: m }),
        body: JSON.stringify({
          projectId: c
        })
      }
    ).then(async (y) => ee(y, c).then((f) => !0));
  }
  async function a(c, m, y) {
    const f = await r();
    return fetch(
      t + `/projects/${c}/firestore_security_rules`,
      {
        method: "PATCH",
        headers: Y({ firebaseAccessToken: f, googleAccessToken: m, serviceAccount: y })
      }
    ).then(async (E) => ee(E, c).then((T) => !0));
  }
  async function l(c, m) {
    const y = await r(), f = {
      ...m,
      roles: m.roles.map((E) => E.id),
      updated_on: /* @__PURE__ */ new Date()
    };
    return fetch(
      t + "/projects/" + c + "/users",
      {
        method: "POST",
        headers: Y({ firebaseAccessToken: y }),
        body: JSON.stringify(f)
      }
    ).then((E) => ee(E, c));
  }
  async function s(c, m, y) {
    const f = await r(), E = {
      ...y,
      roles: y.roles.map((T) => T.id),
      updated_on: /* @__PURE__ */ new Date()
    };
    return console.log("Updating user", E), fetch(
      t + "/projects/" + c + "/users/" + m,
      {
        method: "PATCH",
        headers: Y({ firebaseAccessToken: f }),
        body: JSON.stringify(E)
      }
    ).then((T) => ee(T, c));
  }
  async function d(c, m) {
    const y = await r();
    return fetch(
      t + "/projects/" + c + "/users/" + m,
      {
        method: "DELETE",
        headers: Y({ firebaseAccessToken: y })
      }
    ).then((f) => ee(f, c));
  }
  async function p(c, m, y, f = 10) {
    if (Se[c])
      return Se[c];
    const E = await r();
    async function T() {
      return await new Promise((C) => setTimeout(C, 5e3)), console.debug("Retrying getRootCollections", f), p(c, m, y, f - 1);
    }
    return fetch(
      t + "/projects/" + c + "/firestore_root_collections",
      {
        method: "GET",
        headers: Y({ firebaseAccessToken: E, googleAccessToken: m, serviceAccount: y })
      }
    ).then(async (C) => {
      if (C.status >= 300)
        return await T();
      const P = await ee(C, c);
      return Se[c] = P, P;
    }).catch(async (C) => {
      if (f > 0)
        return await T();
      throw C;
    });
  }
  async function u(c, m) {
    const y = await r();
    return fetch(
      t + "/projects/" + m + "/service_accounts",
      {
        method: "POST",
        headers: Y({ firebaseAccessToken: y, googleAccessToken: c })
      }
    ).then(async (f) => {
      if (f.status === 409)
        throw Error("The service account already exists for this project.");
      return (await f.json()).user;
    });
  }
  async function h(c) {
    const m = await r();
    return fetch(
      t + "/projects/" + c + "/delegated_login",
      {
        method: "POST",
        headers: Y({ firebaseAccessToken: m }),
        body: JSON.stringify({
          projectId: c
        })
      }
    ).then(async (y) => {
      const f = await y.json();
      if (!y.ok)
        throw new Le(f.message, f.code, c, f.data);
      return f.data;
    });
  }
  async function S(c) {
    const m = await r();
    return fetch(
      `${t}/customer/stripe_portal_link?return_url=${encodeURIComponent(window.location.href)}&project_id=${c}`,
      {
        method: "GET",
        headers: Y({ firebaseAccessToken: m })
      }
    ).then(async (y) => (await y.json()).url);
  }
  async function v(c, m) {
    return `${t}/projects/${c}/app_config/${m}/${await r()}/remoteEntry.js`;
  }
  return {
    createNewFireCMSProject: o,
    createFirebaseWebapp: n,
    addSecurityRules: a,
    createServiceAccount: u,
    createNewUser: l,
    updateUser: s,
    deleteUser: d,
    getRootCollections: p,
    doDelegatedLogin: h,
    getStripePortalLink: S,
    getRemoteConfigUrl: v
  };
}
function Y({
  firebaseAccessToken: t,
  googleAccessToken: r,
  serviceAccount: o
}) {
  const n = {
    "Content-Type": "application/json"
  };
  return t && (n.Authorization = `Bearer ${t}`), r && (n["x-admin-authorization"] = `Bearer ${r}`), o && (n["x-admin-service-account"] = `Bearer ${btoa(JSON.stringify(o))}`), n;
}
const He = [
  "https://www.googleapis.com/auth/cloud-platform"
];
function Mr({
  backendApiHost: t,
  backendFirebaseApp: r,
  onUserChange: o
}) {
  const [n, a] = x(void 0), [l, s] = x(!0), [d, p] = x(Ir()), [u, h] = x(), [S, v] = x(!1), [c, m] = x(), [y, f] = x(!0), [E, T] = x(!1), [C, P] = x(), [F, O] = x(), U = ve(), D = j(async (w) => {
    o?.(w), a(w);
  }, []);
  k(() => {
    E && c && Promise.all(c.map((w) => R(w))).then((w) => {
      O(w.filter(Boolean));
    });
  }, [c, E]), k(() => {
    r && (U.current = G(r));
  }, [r]), k(() => {
    if (!r)
      return;
    const w = pe(r);
    return D(w.currentUser), kt(
      w,
      (_) => {
        s(!1), D(_);
      },
      (_) => h(_)
    );
  }, [r, D]), k(() => {
    const w = U.current;
    if (w) {
      if (!n) {
        m(void 0), f(!1), T(!1);
        return;
      }
      return Q(
        J(w, "users", n.uid),
        {
          next: (_) => {
            const M = _.get("projects") ?? [];
            P(void 0), m(M), T(!0), f(!1);
          },
          error: (_) => {
            P(_), f(!1);
          }
        }
      );
    }
  }, [n]);
  const $ = j((w) => {
    if (!r)
      throw Error("useBuildFireCMSBackend googleLogin error");
    const _ = new Me();
    _.setCustomParameters({
      access_type: "offline"
    }), w && He.forEach((H) => _.addScope(H));
    const M = pe(r);
    return Rt(M, _).then((H) => {
      if (w) {
        const W = JSON.parse(H._tokenResponse.rawUserInfo).granted_scopes.split(" ");
        if (w && !W.includes(He[0]))
          v(!0);
        else {
          const me = Me.credentialFromResult(H);
          p(me), Te(me), v(!1);
        }
      }
      return H.user;
    }).catch((H) => (h(H), null));
  }, [r]), B = j(() => {
    const w = pe(r);
    Lr(), At(w).then((_) => {
      a(null), p(null), h(null), Te(null);
    });
  }, [r]), N = j(() => {
    if (!n)
      throw Error("Trying to get Firebase token ");
    return n.getIdToken();
  }, [n]), R = j((w) => {
    const _ = U.current;
    if (!_)
      throw Error("useFireCMSProjectsRepository error");
    return Ne(J(_, "projects", w)).then((M) => M.exists() ? { projectId: M.id, ...M.data() } : null).catch((M) => (console.error("Error getting project:", M), null));
  }, []), b = Or(t, N);
  return {
    backendApiHost: t,
    user: n ?? null,
    signOut: B,
    googleLogin: $,
    getBackendAuthToken: N,
    googleCredential: d,
    availableProjectIds: c,
    availableProjectsLoaded: E,
    availableProjectsLoading: y,
    availableProjectsError: C,
    permissionsNotGrantedError: S,
    authLoading: l,
    authProviderError: u,
    backendFirebaseApp: r,
    backendUid: n?.uid,
    projectsApi: b,
    getProject: R,
    projects: F
  };
}
function Te(t) {
  if (!t) {
    localStorage.removeItem("googleCredential");
    return;
  }
  const r = JSON.stringify({
    created_on: /* @__PURE__ */ new Date(),
    credential: t.toJSON()
  });
  localStorage.setItem("googleCredential", r);
}
function Ir() {
  try {
    const t = localStorage.getItem("googleCredential");
    if (!t)
      return null;
    const r = JSON.parse(t), o = Dt.fromJSON(r.credential), n = new Date(r.created_on);
    return (/* @__PURE__ */ new Date()).getTime() - n.getTime() > 1e3 * 60 * 60 ? (console.debug("Google credential expired credential expired"), Te(null), null) : o;
  } catch (t) {
    return console.error(t), null;
  }
}
function zr({
  backendFirebaseApp: t,
  projectId: r,
  projectsApi: o,
  usersLimit: n,
  canEditRoles: a,
  fireCMSBackend: l
}) {
  const s = r ? `projects/${r}` : void 0, d = ve(), [p, u] = z.useState(!0), [h, S] = z.useState(!0), [v, c] = z.useState([]), [m, y] = z.useState([]), f = m.map((b) => ({
    ...b,
    roles: v.filter((w) => b.roles.includes(w.id))
  })), [E, T] = z.useState(), [C, P] = z.useState();
  k(() => {
    t && (d.current = G(t));
  }, [t]), k(() => {
    const b = d.current;
    if (!(!b || !s))
      return Q(
        ae(b, s, "roles"),
        {
          next: (w) => {
            T(void 0);
            try {
              const _ = Vr(w.docs);
              c(_);
            } catch (_) {
              T(_);
            }
            u(!1);
          },
          error: (w) => {
            T(w), u(!1);
          }
        }
      );
  }, [s]), k(() => {
    const b = d.current;
    if (!(!b || !s))
      return Q(
        ae(b, s, "users"),
        {
          next: (w) => {
            P(void 0);
            try {
              const _ = $r(w.docs);
              y(_);
            } catch (_) {
              P(_);
            }
            S(!1);
          },
          error: (w) => {
            P(w), S(!1);
          }
        }
      );
  }, [s]);
  const F = j(async (b) => {
    if (!d.current || !s)
      throw Error("useFirestoreConfigurationPersistence Firestore not initialised");
    console.debug("Persisting", b);
    const {
      uid: _,
      ...M
    } = b;
    return _ ? o.updateUser(r, _, b) : o.createNewUser(r, b);
  }, [s, r]), O = j((b) => {
    const w = d.current;
    if (!w || !s)
      throw Error("useFirestoreConfigurationPersistence Firestore not initialised");
    console.debug("Persisting", b);
    const {
      id: _,
      ...M
    } = b, H = J(w, s, "roles", _);
    return se(H, M, { merge: !0 });
  }, [s]), U = j(async (b) => {
    if (!d.current || !s)
      throw Error("useFirestoreConfigurationPersistence Firestore not initialised");
    console.debug("Deleting", b);
    const { uid: _ } = b;
    return o.deleteUser(r, _);
  }, [s]), D = j((b) => {
    const w = d.current;
    if (!w || !s)
      throw Error("useFirestoreConfigurationPersistence Firestore not initialised");
    console.debug("Deleting", b);
    const { id: _ } = b, M = J(w, s, "roles", _);
    return St(M);
  }, [s]), $ = f.find((b) => b.email.toLowerCase() === l.user?.email?.toLowerCase()), B = j(({
    collection: b
  }) => Ge({
    collection: b,
    user: $ ?? null
  }), [$?.uid]), N = f.map((b) => b.uid), R = j((b) => {
    if (!f)
      throw Error("Users not loaded");
    return f.find((_) => _.email?.toLowerCase() === b?.email?.toLowerCase())?.roles;
  }, [N]);
  return {
    allowDefaultRolesCreation: !1,
    includeCollectionConfigPermissions: !0,
    loading: p || h,
    roles: v,
    users: f,
    saveUser: F,
    saveRole: O,
    defineRolesFor: R,
    deleteUser: U,
    deleteRole: D,
    usersLimit: n,
    canEditRoles: a === void 0 ? !0 : a,
    collectionPermissions: B
  };
}
const $r = (t) => t.map((r) => ({
  uid: r.id,
  ...r.data(),
  created_on: r.data()?.created_on?.toDate(),
  updated_on: r.data()?.updated_on?.toDate()
})), Vr = (t) => t.map((r) => ({
  id: r.id,
  ...r.data()
}));
function Br({
  hovered: t,
  drawerOpen: r,
  closeDrawer: o
}) {
  const n = nt(), a = he(), l = ot(), { user: s } = We(), [d, p] = z.useState(!1), u = t && !r && !d;
  if (!a.topLevelNavigation)
    throw Error("Navigation not ready in Drawer");
  const {
    navigationEntries: h,
    groups: S
  } = a.topLevelNavigation, v = j((c) => {
    if (!r)
      return /* @__PURE__ */ e("div", { style: { height: 16 } });
    const m = c && st.includes(c), y = l.configPermissions({ user: s }).createCollections && !m;
    return /* @__PURE__ */ i("div", { className: "flex flex-row items-center pt-8 pl-8 pr-0 pb-2", children: [
      /* @__PURE__ */ e(
        g,
        {
          variant: "caption",
          color: "secondary",
          className: "flex-grow font-medium",
          children: c ? c.toUpperCase() : "Views".toUpperCase()
        }
      ),
      y && /* @__PURE__ */ e(
        Ie,
        {
          title: c ? `Create new collection in ${c}` : "Create new collection",
          children: /* @__PURE__ */ e(
            V,
            {
              size: "small",
              variant: "text",
              onClick: () => {
                l?.createCollection({
                  initialValues: {
                    group: c
                  },
                  parentCollectionIds: [],
                  redirect: !0,
                  sourceClick: "drawer_new_collection"
                });
              },
              children: /* @__PURE__ */ e(Qe, { size: "small" })
            }
          )
        }
      )
    ] });
  }, [l, r]);
  return /* @__PURE__ */ i(A, { children: [
    /* @__PURE__ */ e("div", { className: "flex-grow overflow-scroll no-scrollbar", children: S.map((c) => /* @__PURE__ */ i(
      z.Fragment,
      {
        children: [
          v(c),
          Object.values(h).filter((m) => m.group === c).map((m, y) => /* @__PURE__ */ e(
            Ot,
            {
              icon: /* @__PURE__ */ e(Ee, { collectionOrView: m.collection ?? m.view }),
              tooltipsOpen: u,
              drawerOpen: r,
              url: m.url,
              name: m.name
            },
            `navigation_${y}`
          ))
        ]
      },
      `drawer_group_${c}`
    )) }),
    /* @__PURE__ */ e(
      Kt,
      {
        open: d,
        onOpenChange: p,
        trigger: /* @__PURE__ */ i(
          ke,
          {
            shape: "square",
            className: "m-4 text-gray-900 dark:text-white w-fit",
            children: [
              /* @__PURE__ */ e(
                Ie,
                {
                  title: "Admin",
                  open: u,
                  side: "right",
                  sideOffset: 28,
                  children: /* @__PURE__ */ e(Xt, {})
                }
              ),
              r && /* @__PURE__ */ e(
                "div",
                {
                  className: re(
                    r ? "opacity-100" : "opacity-0 hidden",
                    "mx-4 font-inherit text-inherit"
                  ),
                  children: "ADMIN"
                }
              )
            ]
          }
        ),
        children: at.map((c, m) => /* @__PURE__ */ i(
          Zt,
          {
            onClick: (y) => {
              y.preventDefault(), n(c.path);
            },
            children: [
              /* @__PURE__ */ e(Ee, { collectionOrView: c }),
              c.name
            ]
          },
          `navigation_${m}`
        ))
      }
    )
  ] });
}
function Hr({ projectId: t, context: r }) {
  return /* @__PURE__ */ i("div", { className: "flex flex-col gap-1 p-1", children: [
    /* @__PURE__ */ e(g, { variant: "h6", children: "Subscription required" }),
    /* @__PURE__ */ e(g, { children: "You have finished your free usage quota." }),
    /* @__PURE__ */ e(g, { children: "Please upgrade plans to continue using this feature in this project" }),
    /* @__PURE__ */ e(
      V,
      {
        color: "primary",
        component: "a",
        rel: "noopener noreferrer",
        target: "_blank",
        href: `/p/${t}/settings`,
        children: "Upgrade plan"
      }
    )
  ] });
}
function Jr({}) {
  return /* @__PURE__ */ i(A, { children: [
    /* @__PURE__ */ e(g, { variant: "h4", children: "Security rules" }),
    /* @__PURE__ */ i(g, { children: [
      "FireCMS uses Firebase security rules to restrict access to data. When creating a new user through FireCMS, the custom claim ",
      /* @__PURE__ */ e(
        g,
        {
          className: "inline-block bg-white dark:bg-gray-950 px-1 rounded-md",
          component: "pre",
          children: "fireCMSUser"
        }
      ),
      " is added to the user in the client project. By adding the following security rules to your project, you ensure FireCMS users can access data through FireCMS."
    ] }),
    /* @__PURE__ */ e(Re, { children: /* @__PURE__ */ e(
      g,
      {
        component: "pre",
        className: "m-0 p-4 text-sm font-mono",
        children: `match /{document=**} {
    allow read, write: if request.auth.token.fireCMSUser;
}`
      }
    ) }),
    /* @__PURE__ */ i(g, { variant: "caption", children: [
      "This rules restrict access to data to FireCMS users only, but does ",
      /* @__PURE__ */ e("strong", { children: "not" }),
      " enforce the permissions at the database level."
    ] }),
    /* @__PURE__ */ e(g, { variant: "caption", children: "The permissions are enforced in the frontend though, which will work fine for most users. If you need to enforce the permissions at the database level, you can modify these security rules yourself to suit your needs. The roles assigned to a user are set as a custom claim in the Firebase auth token, so you can use them in your security rules." })
  ] });
}
function Yr({
  error: t,
  fireCMSBackend: r,
  onFixed: o,
  onRetry: n
}) {
  const {
    code: a,
    message: l,
    projectId: s
  } = t;
  return a === "service-account-missing" && s && r ? /* @__PURE__ */ e(
    Gr,
    {
      projectId: s,
      fireCMSBackend: r,
      onFixed: o
    }
  ) : a === "user-has-to-accept-googles-terms-of-service" && s ? /* @__PURE__ */ e(
    Wr,
    {
      projectId: s,
      onFixed: o
    }
  ) : a === "user-has-no-previous-firebase-projects" && s ? /* @__PURE__ */ e(
    qr,
    {
      projectId: s,
      onFixed: o
    }
  ) : a === "firecms-user-not-found" ? /* @__PURE__ */ i("div", { className: "flex flex-col items-center space-y-2 py-4", children: [
    /* @__PURE__ */ e(g, { children: "The user trying to log in is not registered in the client project." }),
    /* @__PURE__ */ i(g, { children: [
      "Make sure the user exists in the client project and try again. If the problem persists, reach us at ",
      /* @__PURE__ */ e(
        "a",
        {
          href: "mailto:hello@firecms.co?subject=FireCMS%20login%20error",
          rel: "noopener noreferrer",
          target: "_blank",
          children: "hello@firecms.co "
        }
      ),
      ", or in our ",
      /* @__PURE__ */ e(
        "a",
        {
          rel: "noopener noreferrer",
          target: "_blank",
          href: "https://discord.gg/fxy7xsQm3m",
          children: "Discord channel"
        }
      ),
      "."
    ] })
  ] }) : a === "service-account-missing-permissions" ? /* @__PURE__ */ e(Qr, { missingPermissions: t.data?.missingPermissions }) : /* @__PURE__ */ i("div", { className: "flex flex-col items-center space-y-2 py-4", children: [
    /* @__PURE__ */ e(g, { className: "text-center text-red-300", children: l }),
    n && /* @__PURE__ */ e(
      V,
      {
        variant: "outlined",
        color: "error",
        onClick: () => n(),
        children: "Retry"
      }
    )
  ] });
}
function Gr({
  fireCMSBackend: t,
  projectId: r,
  onFixed: o
}) {
  const { projectsApi: n } = oe(), a = Ue(), [l, s] = x(!1), [d, p] = x(!1), u = async () => {
    if (!t.googleCredential?.accessToken)
      throw new Error("SassMissingServiceAccountErrorView: No access token found");
    return s(!0), n.createServiceAccount(t.googleCredential.accessToken, r).finally(() => s(!1));
  };
  return k(() => {
    d && t.googleCredential?.accessToken && (u().then(() => {
      a.open({
        type: "success",
        message: "Service account created successfully"
      }), o && o();
    }).catch((S) => {
      a.open({
        type: "error",
        message: "Service account creation error: " + S.message
      });
    }), p(!1));
  }, [d, t.googleCredential?.accessToken]), /* @__PURE__ */ i(
    "div",
    {
      className: "flex flex-col items-center space-y-2 py-4",
      children: [
        /* @__PURE__ */ e(g, { color: "error", children: "Service account missing" }),
        /* @__PURE__ */ e(
          fe,
          {
            variant: "outlined",
            color: "error",
            onClick: async () => {
              t.googleCredential?.accessToken ? await u() : (p(!0), t.googleLogin(!0));
            },
            loading: l,
            startIcon: /* @__PURE__ */ e(Ae, {}),
            children: "Fix"
          }
        )
      ]
    }
  );
}
function Wr({
  projectId: t,
  onFixed: r
}) {
  return /* @__PURE__ */ i(
    "div",
    {
      className: "flex flex-col items-center space-y-2 py-4",
      children: [
        /* @__PURE__ */ e(g, { color: "error", children: "You need to accept Google's terms of service before you can use this service." }),
        /* @__PURE__ */ i(g, { color: "error", children: [
          "You can do so by visiting the following link:",
          /* @__PURE__ */ e(
            "a",
            {
              rel: "noopener noreferrer",
              target: "_blank",
              href: "https://console.cloud.google.com/welcome?project=" + t,
              children: `https://console.cloud.google.com/welcome?project=${t}`
            }
          )
        ] }),
        /* @__PURE__ */ e(
          V,
          {
            variant: "outlined",
            color: "error",
            onClick: r,
            startIcon: /* @__PURE__ */ e(Ae, {}),
            children: "I have accepted the terms"
          }
        )
      ]
    }
  );
}
function qr({
  projectId: t,
  onFixed: r
}) {
  return /* @__PURE__ */ i(
    "div",
    {
      className: "flex flex-col items-center space-y-2 py-4",
      children: [
        /* @__PURE__ */ e(g, { color: "error", children: "You need to accept Firebase's terms of service before you can use this service." }),
        /* @__PURE__ */ i(g, { color: "error", children: [
          "You can do so by visiting the following link:",
          /* @__PURE__ */ e(
            "a",
            {
              rel: "noopener noreferrer",
              target: "_blank",
              href: "https://console.firebase.google.com/",
              children: "https://console.firebase.google.com/"
            }
          )
        ] }),
        /* @__PURE__ */ e(
          V,
          {
            variant: "outlined",
            color: "error",
            onClick: r,
            startIcon: /* @__PURE__ */ e(Ae, {}),
            children: "I have accepted the terms"
          }
        )
      ]
    }
  );
}
function Qr(t) {
  return /* @__PURE__ */ i(
    "div",
    {
      className: "flex flex-col space-y-2 py-4",
      children: [
        /* @__PURE__ */ e(g, { color: "error", children: "Missing permissions" }),
        t.missingPermissions && /* @__PURE__ */ i(A, { children: [
          /* @__PURE__ */ e(g, { variant: "body2", children: "This service account is missing the following permissions:" }),
          /* @__PURE__ */ e(g, { variant: "caption", component: "ul", children: t.missingPermissions.map((r) => /* @__PURE__ */ e("li", { children: /* @__PURE__ */ e("code", { children: r }) }, r)) })
        ] })
      ]
    }
  );
}
function te(t) {
  return t.currency === "usd" ? new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: t.currency
  }).format(t.unit_amount / 100) + " user/" + t.interval : new Intl.NumberFormat("es-ES", {
    style: "currency",
    currency: t.currency
  }).format(t.unit_amount / 100) + " user/" + t.interval;
}
function ut(t) {
  return t.status === "active" ? "Active" : t.status === "trialing" ? "Trialing" : t.status === "past_due" ? "Past due" : t.status === "canceled" ? "Canceled" : t.status === "unpaid" ? "Unpaid" : t.status === "incomplete" ? "Incomplete" : t.status === "incomplete_expired" ? "Incomplete expired" : "Unknown";
}
function ft(t) {
  switch (t) {
    case "free":
      return "Free";
    case "cloud_plus":
      return "Plus";
    case "cloud_pro":
      return "Pro";
  }
}
function ht({ subscriptionPlan: t }) {
  const r = ft(t);
  return /* @__PURE__ */ e(
    K,
    {
      className: "uppercase font-medium",
      colorScheme: "tealDarker",
      size: "small",
      children: r + " plan"
    }
  );
}
function mt({
  product: t,
  includePriceSelect: r = !0,
  includePriceLabel: o = !0,
  largePriceLabel: n = !1,
  subscribe: a,
  projectId: l
}) {
  const [s, d] = x(), [p, u] = x(), h = t.prices;
  if (k(() => {
    h.length > 0 && u(h[0]);
  }, [h]), t.metadata.type !== "cloud_pro" && t.metadata.type !== "cloud_plus")
    throw new Error("Error: Unmapped product type in ProductView");
  const S = ft(t.metadata.type), v = h?.length > 1 ? /* @__PURE__ */ e(A, { children: /* @__PURE__ */ e(
    er,
    {
      size: "small",
      invisible: !0,
      padding: !1,
      onChange: (f) => {
        u(h.find((E) => E.id === f.target.value));
      },
      position: "item-aligned",
      renderValue: (f) => {
        const E = h.find((T) => T.id === f);
        return E ? n ? /* @__PURE__ */ e(
          "span",
          {
            className: "ml-4 mb-4 text-2xl font-bold text-primary text-center my-8",
            children: te(E)
          }
        ) : /* @__PURE__ */ e(K, { children: E ? te(E) : "" }) : null;
      },
      value: p?.id ?? "",
      children: h && h.map(
        (f) => /* @__PURE__ */ e(tr, { value: f.id, children: te(f) }, f.id)
      )
    }
  ) }) : h ? n ? /* @__PURE__ */ e(
    "span",
    {
      className: "ml-4 mb-4 text-2xl font-bold text-primary text-center my-8",
      children: te(h[0])
    }
  ) : /* @__PURE__ */ e(
    K,
    {
      size: "small",
      children: te(h[0])
    }
  ) : null, [c, m] = x(!1);
  return /* @__PURE__ */ i(A, { children: [
    r && /* @__PURE__ */ i(A, { children: [
      o && /* @__PURE__ */ i("div", { className: "my-2 flex items-center gap-2", children: [
        "You can upgrade your project for ",
        v
      ] }),
      !o && v
    ] }),
    /* @__PURE__ */ i(
      fe,
      {
        variant: "filled",
        loading: c,
        onClick: () => {
          if (!(!l || p === void 0))
            return m(!0), a(
              l,
              p,
              (f, E) => {
                !f && !E || (E && d(E), f && typeof window < "u" && window.open(f, "_blank"), m(!1));
              },
              t.metadata.type
            );
        },
        startIcon: /* @__PURE__ */ e(Ke, {}),
        children: [
          "Upgrade to ",
          S
        ]
      }
    ),
    s && /* @__PURE__ */ e(De, { color: "error", className: "my-4", children: s.message })
  ] });
}
function gt({
  includePriceSelect: t,
  includePriceLabel: r,
  largePriceLabel: o
}) {
  const { backendFirebaseApp: n } = oe(), {
    subscriptionPlan: a,
    projectId: l
  } = X();
  if (!a)
    throw new Error("No subscription plan");
  const {
    products: s,
    subscribe: d,
    getSubscriptionsForProject: p
  } = dt({
    firebaseApp: n
  }), u = s?.find((h) => h.metadata?.type === "cloud_plus");
  return u ? /* @__PURE__ */ e(
    mt,
    {
      includePriceSelect: t,
      includePriceLabel: r,
      largePriceLabel: o,
      product: u,
      projectId: l,
      subscribe: d
    }
  ) : /* @__PURE__ */ e(
    fe,
    {
      variant: "filled",
      loading: !0,
      startIcon: /* @__PURE__ */ e(Ke, {}),
      children: "Upgrade to PLUS"
    }
  );
}
function Kr({
  open: t,
  onClose: r
}) {
  return he(), nt(), /* @__PURE__ */ i(
    Xe,
    {
      fullScreen: !0,
      open: t,
      onOpenChange: (o) => o ? void 0 : r(),
      children: [
        /* @__PURE__ */ e(Ze, { fullHeight: !0, children: /* @__PURE__ */ e(Xr, {}) }),
        /* @__PURE__ */ e(
          ke,
          {
            className: "absolute top-4 right-4",
            onClick: r,
            children: /* @__PURE__ */ e(et, {})
          }
        )
      ]
    }
  );
}
function Xr() {
  return /* @__PURE__ */ e("div", { className: "overflow-auto my-auto", children: /* @__PURE__ */ i(rr, { className: "flex flex-col gap-4 p-8 m-auto", children: [
    /* @__PURE__ */ e("h2", { className: "text-3xl md:text-4xl font-bold my-4 text-center", children: "Full no-code/low-code solution" }),
    /* @__PURE__ */ i("p", { children: [
      /* @__PURE__ */ e("strong", { children: "FireCMS Cloud" }),
      " offers a complete, end-to-end solution for businesses that require the highest level of support and security. With dedicated hosting, advanced features, and expert support, you will have everything you need to take your project to the next level."
    ] }),
    /* @__PURE__ */ i(
      "div",
      {
        className: "flex flex-col items-center lg:grid lg:grid-cols-3 gap-4 mt-8 w-full mx-auto",
        children: [
          /* @__PURE__ */ i(
            "div",
            {
              className: "h-full max-w-sm p-6 border border-solid border-gray-200 rounded-lg shadow dark:border-gray-700 flex flex-col",
              children: [
                /* @__PURE__ */ e("h3", { className: "text-2xl md:text-4xl font-bold text-center text-gray-800 dark:text-gray-200 uppercase my-4", children: "Free" }),
                /* @__PURE__ */ i("div", { className: "grow", children: [
                  /* @__PURE__ */ e("p", { className: "text-lg mb-4 text-center font-semibold", children: "Try FireCMS and upgrade to a paid plan when you need more features." }),
                  /* @__PURE__ */ i("ul", { children: [
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Unlimited projects" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Unlimited collections" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "All available form fields" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Schema editor and data inference" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Advanced data import and export" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Default roles" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "3 users" })
                  ] })
                ] }),
                /* @__PURE__ */ e("div", { className: "text-center mt-4 text-gray-600 dark:text-gray-400 w-full", children: /* @__PURE__ */ e("span", { className: "text-2xl font-bold ", children: "â‚¬0 user/month" }) })
              ]
            }
          ),
          /* @__PURE__ */ i(
            "div",
            {
              className: "h-full max-w-sm p-6 rounded-lg flex flex-col outline-none ring-2 ring-primary ring-opacity-75 ring-offset-2 ring-offset-transparent",
              children: [
                /* @__PURE__ */ e("h3", { className: "text-2xl md:text-4xl font-bold text-center dark:text-primary text-primary uppercase my-4", children: "Plus" }),
                /* @__PURE__ */ i("div", { className: "grow", children: [
                  /* @__PURE__ */ e("p", { className: "text-lg mb-4 text-center font-semibold", children: "Perfect for startups" }),
                  /* @__PURE__ */ i("ul", { children: [
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Everything in the free tier" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Local text search" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Custom fields and custom views" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Unlimited users and roles" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Unlimited data export" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Theme and logo customization" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Custom user roles" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "GPT-4 content generation" })
                  ] })
                ] }),
                /* @__PURE__ */ e("div", { className: "flex flex-col gap-4 items-center justify-center mt-8", children: /* @__PURE__ */ e(
                  gt,
                  {
                    includePriceSelect: !0,
                    includePriceLabel: !1,
                    largePriceLabel: !0
                  }
                ) })
              ]
            }
          ),
          /* @__PURE__ */ i(
            "div",
            {
              className: "h-full max-w-sm p-6 border border-solid border-gray-200 rounded-lg shadow dark:border-gray-700 flex flex-col",
              children: [
                /* @__PURE__ */ e("h3", { className: "text-2xl md:text-4xl font-bold text-center text-gray-800 dark:text-gray-200 uppercase my-4", children: "Pro" }),
                /* @__PURE__ */ i("div", { className: "grow", children: [
                  /* @__PURE__ */ e("p", { className: "text-lg mb-4 text-center font-semibold", children: "Perfect for agencies and larger organizations" }),
                  /* @__PURE__ */ i("ul", { children: [
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Everything in PLUS" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Self-hosted" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "SAML SSO" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Custom domain" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Full CMS components customization" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Priority support" }),
                    /* @__PURE__ */ e("li", { className: "ml-8 list-disc", children: "Roadmap prioritization" })
                  ] })
                ] }),
                /* @__PURE__ */ e("div", { className: "text-center mt-4 text-primary w-full", children: /* @__PURE__ */ e(
                  V,
                  {
                    component: "a",
                    size: "large",
                    variant: "outlined",
                    href: "https://firecms.co/pro",
                    rel: "noopener noreferrer",
                    target: "_blank",
                    children: "More info"
                  }
                ) })
              ]
            }
          )
        ]
      }
    )
  ] }) });
}
function Zr() {
  return /* @__PURE__ */ i(g, { variant: "caption", className: "mt-4", children: [
    "You will be redirected to ",
    /* @__PURE__ */ e("b", { children: "Stripe" }),
    " to manage the subscription. You will be charged every month per the maximum concurrent of users. You can cancel your subscription at any time, and enjoy the benefits until the end of the current billing period."
  ] });
}
function eo({ uid: t }) {
  const { backendFirebaseApp: r } = oe(), {
    subscriptionPlan: o,
    projectId: n
  } = X();
  if (!o)
    throw new Error("No subscription plan");
  je("Plans");
  const {
    products: a,
    subscribe: l,
    getSubscriptionsForProject: s
  } = dt({
    firebaseApp: r
  }), d = s(n), p = d === void 0 || a === void 0, h = (a ?? []).filter((v) => v.metadata?.type === "cloud_plus" || v.metadata?.type === "cloud_pro").find((v) => v.metadata?.type === "cloud_plus"), S = d.find((v) => v.product.metadata?.type === "cloud_plus");
  return /* @__PURE__ */ i("div", { className: "relative", children: [
    p && /* @__PURE__ */ e("div", { className: "absolute w-full h-full flex items-center justify-center", children: /* @__PURE__ */ e(tt, {}) }),
    /* @__PURE__ */ i("div", { className: re("grid grid-cols-12 gap-4 items-center", p ? "collapse" : ""), children: [
      /* @__PURE__ */ i("div", { className: "col-span-12 md:col-span-7 flex flex-col gap-2", children: [
        /* @__PURE__ */ e(g, { variant: "h4", className: "mt-4 mb-2", children: "Subscription Plan" }),
        /* @__PURE__ */ i(
          g,
          {
            variant: "subtitle1",
            className: "my-2",
            children: [
              "This project is on the ",
              /* @__PURE__ */ e(ht, { subscriptionPlan: o })
            ]
          }
        ),
        o !== "cloud_plus" && h && /* @__PURE__ */ e(
          mt,
          {
            product: h,
            projectId: n,
            subscribe: l
          }
        ),
        o === "cloud_plus" && S && /* @__PURE__ */ e(to, { subscription: S }),
        /* @__PURE__ */ e(Zr, {})
      ] }),
      /* @__PURE__ */ e("div", { className: re("col-span-12 md:col-span-5"), children: /* @__PURE__ */ i(
        or,
        {
          className: "p-6 bg-amber-200 dark:bg-amber-700 border-amber-300 dark:border-amber-800 flex flex-col gap-4",
          children: [
            o === "free" && /* @__PURE__ */ i("div", { children: [
              "Upgrade to the ",
              /* @__PURE__ */ e("b", { children: "PLUS" }),
              " plan to enable the following features:"
            ] }),
            o === "cloud_plus" && /* @__PURE__ */ i("div", { children: [
              "By being on the ",
              /* @__PURE__ */ e("b", { children: "PLUS" }),
              " plan you are enjoying the following features:"
            ] }),
            /* @__PURE__ */ i("ul", { className: "px-2 text-base", children: [
              /* @__PURE__ */ i("li", { className: "flex gap-4 items-center py-0.5", children: [
                /* @__PURE__ */ e(Z, { size: "small" }),
                "Local text search"
              ] }),
              /* @__PURE__ */ i("li", { className: "flex gap-4 items-center py-0.5", children: [
                /* @__PURE__ */ e(Z, { size: "small" }),
                "Unlimited users and roles"
              ] }),
              /* @__PURE__ */ i("li", { className: "flex gap-4 items-center py-0.5", children: [
                /* @__PURE__ */ e(Z, { size: "small" }),
                "Theme and logo customization"
              ] }),
              /* @__PURE__ */ i("li", { className: "flex gap-4 items-center py-0.5", children: [
                /* @__PURE__ */ e(Z, { size: "small" }),
                "Custom fields and custom views"
              ] }),
              /* @__PURE__ */ i("li", { className: "flex gap-4 items-center py-0.5", children: [
                /* @__PURE__ */ e(Z, { size: "small" }),
                "Custom user roles"
              ] }),
              /* @__PURE__ */ i("li", { className: "flex gap-4 items-center py-0.5", children: [
                /* @__PURE__ */ e(Z, { size: "small" }),
                "GPT-4 content generation"
              ] }),
              /* @__PURE__ */ i("li", { className: "flex gap-4 items-center py-0.5", children: [
                /* @__PURE__ */ e(Z, { size: "small" }),
                "Unlimited data export"
              ] })
            ] })
          ]
        }
      ) })
    ] })
  ] });
}
function to({
  subscription: t
}) {
  const {
    getBackendAuthToken: r,
    projectsApi: o
  } = oe(), n = ut(t), [a, l] = x(void 0);
  return k(() => {
    a || o.getStripePortalLink(t.metadata.projectId).then(l);
  }, []), /* @__PURE__ */ e(
    "div",
    {
      className: "mt-2 mb-2",
      children: /* @__PURE__ */ i("div", { children: [
        /* @__PURE__ */ i(A, { children: [
          "The subscription is ",
          /* @__PURE__ */ i(
            K,
            {
              className: "inline",
              size: "tiny",
              colorScheme: n === "Active" ? "greenDark" : "orangeDark",
              children: [
                n,
                " "
              ]
            }
          ),
          "."
        ] }),
        t.current_period_end && /* @__PURE__ */ i(A, { children: [
          " The next payment is on ",
          t.current_period_end.toDate().toLocaleDateString(),
          ". "
        ] }),
        /* @__PURE__ */ i(A, { children: [
          "The current price is ",
          /* @__PURE__ */ e(
            K,
            {
              size: "tiny",
              children: te(t.price)
            }
          ),
          "per active user."
        ] }),
        t.cancel_at && /* @__PURE__ */ i(A, { children: [
          " This subscription was ",
          /* @__PURE__ */ e("b", { children: "cancelled" }),
          " and will be active until ",
          t.cancel_at.toDate().toLocaleDateString(),
          ". "
        ] }),
        /* @__PURE__ */ e(
          "a",
          {
            className: " " + t.canceled_at ? void 0 : "text-text-secondary dark:text-text-secondary-dark",
            href: a,
            target: "_blank",
            rel: "noreferrer",
            children: t.canceled_at ? " Renew subscription" : " Cancel subscription"
          }
        )
      ] })
    },
    t.id
  );
}
function ce({ message: t, showForPlans: r, includeCTA: o = !0 }) {
  const { subscriptionPlan: n } = X(), [a, l] = z.useState(!1);
  return !n || r && !r.includes(n) ? null : /* @__PURE__ */ i(A, { children: [
    /* @__PURE__ */ i(
      De,
      {
        color: "info",
        className: "my-4",
        action: o && /* @__PURE__ */ e(
          V,
          {
            className: "dark:!text-white dark:border-white dark:hover:bg-white dark:hover:!text-primary min-w-content",
            variant: "outlined",
            onClick: () => l(!0),
            children: "More info"
          }
        ),
        children: [
          /* @__PURE__ */ i("div", { children: [
            "This project is currently in the ",
            /* @__PURE__ */ e(ht, { subscriptionPlan: n })
          ] }),
          /* @__PURE__ */ e("div", { children: t })
        ]
      }
    ),
    /* @__PURE__ */ e(Kr, { open: a, onClose: () => l(!1) })
  ] });
}
function ro({
  subscription: t
}) {
  const { projectsApi: r } = oe(), o = ut(t), [n, a] = x(void 0);
  return k(() => {
    n || r.getStripePortalLink(t.metadata.projectId).then(a);
  }, []), /* @__PURE__ */ i(
    Re,
    {
      className: "mt-2 mb-2 p-4 -mx-2",
      children: [
        /* @__PURE__ */ i("div", { className: "flex", children: [
          /* @__PURE__ */ e(
            g,
            {
              className: "flex-grow",
              variant: "h6",
              children: t.product.name
            }
          ),
          /* @__PURE__ */ e(
            K,
            {
              size: "small",
              colorScheme: o === "Active" ? "greenDark" : "orangeDark",
              children: o
            }
          )
        ] }),
        /* @__PURE__ */ i("div", { className: "flex mt-2 items-center", children: [
          /* @__PURE__ */ i(
            "div",
            {
              className: "flex-grow flex flex-col items-start gap-2",
              children: [
                /* @__PURE__ */ i(g, { children: [
                  "Project: ",
                  /* @__PURE__ */ e("b", { children: t.metadata.projectId })
                ] }),
                /* @__PURE__ */ e(
                  K,
                  {
                    size: "small",
                    children: te(t.price) + " per active user"
                  }
                ),
                t.cancel_at && /* @__PURE__ */ i(
                  g,
                  {
                    variant: "caption",
                    className: "text-secondary",
                    children: [
                      "This subscription was ",
                      /* @__PURE__ */ e("b", { children: "cancelled" }),
                      " and will be active until ",
                      t.cancel_at.toDate().toLocaleDateString()
                    ]
                  }
                )
              ]
            }
          ),
          /* @__PURE__ */ e(
            V,
            {
              component: "a",
              variant: "text",
              className: " " + t.canceled_at ? void 0 : "text-text-secondary dark:text-text-secondary-dark",
              size: "small",
              href: n,
              target: "_blank",
              rel: "noreferrer",
              children: t.canceled_at ? "Renew" : "Cancel"
            }
          )
        ] })
      ]
    },
    t.id
  );
}
function $o({ activeSubscriptions: t }) {
  return /* @__PURE__ */ i("div", { className: re("my-8 border-t", nr), children: [
    /* @__PURE__ */ e(g, { className: "my-4 mt-8 font-medium", children: "Active Subscriptions".toUpperCase() }),
    t.map((r) => /* @__PURE__ */ e(
      ro,
      {
        subscription: r
      },
      r.id
    ))
  ] });
}
function oo() {
  const { backendUid: t } = oe(), r = X();
  if (je("Project settings"), !t)
    throw new Error("No backendUid in ProjectSettings");
  return /* @__PURE__ */ i(
    de,
    {
      maxWidth: "6xl",
      className: "w-full flex flex-col gap-16 px-4 py-16",
      children: [
        /* @__PURE__ */ e(eo, { uid: t }),
        /* @__PURE__ */ i("div", { className: "flex flex-col gap-2", children: [
          /* @__PURE__ */ e(g, { variant: "h4", className: "mt-4 mb-2", children: "Settings" }),
          /* @__PURE__ */ e(no, {}),
          /* @__PURE__ */ e("div", { className: "col-span-12", children: /* @__PURE__ */ e(
            _e,
            {
              position: "start",
              label: "Enable local text search",
              onValueChange: (o) => r.updateLocalTextSearchEnabled(o),
              value: r.localTextSearchEnabled
            }
          ) }),
          /* @__PURE__ */ e(ao, {})
        ] }),
        /* @__PURE__ */ e("div", { className: "flex flex-col gap-2", children: /* @__PURE__ */ e(Jr, {}) })
      ]
    }
  );
}
function no() {
  const t = X(), [r, o] = x(t.projectName ?? ""), n = Ct(r);
  return k(() => {
    n && t.updateProjectName(n);
  }, [n]), /* @__PURE__ */ e(
    ir,
    {
      value: r,
      label: "Project name",
      onChange: (a) => o(a.target.value),
      onBlur: () => {
        r && t.updateProjectName(r);
      }
    }
  );
}
function io({ onNoSubscriptionPlan: t }) {
  const {
    logo: r,
    canModifyTheme: o,
    uploadLogo: n
  } = X(), a = Ue(), l = async (d) => {
    if (!o) {
      t();
      return;
    }
    d.length && n(d[0]);
  }, s = (d, p) => {
    if (!o)
      t();
    else
      for (const u of d)
        for (const h of u.errors)
          a.open({
            type: "error",
            message: `Error uploading file: ${h.message}`
          });
  };
  return /* @__PURE__ */ i(
    sr,
    {
      accept: { "*/image": [] },
      maxSize: 2048 * 1024,
      onFilesAdded: l,
      onFilesRejected: s,
      uploadDescription: "Drag and drop your logo here",
      children: [
        r && /* @__PURE__ */ e(
          "img",
          {
            className: "w-40 h-40 p-4",
            src: r
          }
        ),
        !r && /* @__PURE__ */ e(
          qe,
          {
            className: "w-40 h-40 p-4"
          }
        )
      ]
    }
  );
}
function so() {
  const [t, r] = x(!0);
  return /* @__PURE__ */ i("div", { className: "p-4 mt-4 flex flex-col items-center gap-2", children: [
    /* @__PURE__ */ e(g, { variant: "label", children: "Sample theme components" }),
    /* @__PURE__ */ i("div", { className: "flex flex-row gap-4 items-center justify-center", children: [
      /* @__PURE__ */ e(V, { children: " Button " }),
      /* @__PURE__ */ e(V, { variant: "outlined", children: " Button " }),
      /* @__PURE__ */ e(ar, { value: t, onValueChange: r }),
      /* @__PURE__ */ e(ze, { checked: t, onCheckedChange: r }),
      /* @__PURE__ */ e(ze, { color: "secondary", checked: t, onCheckedChange: r })
    ] })
  ] });
}
function ao() {
  const [t, r] = x(!1), o = X();
  return /* @__PURE__ */ i("div", { className: "flex flex-col gap-2 mt-4 mb-2", children: [
    /* @__PURE__ */ e(g, { variant: "h4", className: "mt-4 mb-2", children: "Theme" }),
    /* @__PURE__ */ i("div", { className: "grid grid-cols-12 gap-4", children: [
      /* @__PURE__ */ e("div", { className: "col-span-12 md:col-span-6", children: /* @__PURE__ */ e(io, { onNoSubscriptionPlan: () => r(!0) }) }),
      /* @__PURE__ */ e("div", { className: "col-span-12 md:col-span-6", children: /* @__PURE__ */ i(Re, { className: "flex flex-col gap-2 p-4", children: [
        /* @__PURE__ */ i("div", { className: "flex flex-row gap-4 justify-center mt-4", children: [
          /* @__PURE__ */ i("div", { className: "flex flex-row gap-2", children: [
            /* @__PURE__ */ e(
              "input",
              {
                type: "color",
                value: o.primaryColor,
                onChange: (n) => (r(!0), o.updatePrimaryColor(n.target.value))
              }
            ),
            /* @__PURE__ */ e(g, { variant: "subtitle2", children: "Primary color" })
          ] }),
          /* @__PURE__ */ i("div", { className: "flex flex-row gap-2", children: [
            /* @__PURE__ */ e(
              "input",
              {
                type: "color",
                value: o.secondaryColor,
                onChange: (n) => (r(!0), o.updateSecondaryColor(n.target.value))
              }
            ),
            /* @__PURE__ */ e(g, { variant: "subtitle2", children: "Secondary color" })
          ] })
        ] }),
        /* @__PURE__ */ e(so, {})
      ] }) })
    ] }),
    t && /* @__PURE__ */ e(
      ce,
      {
        showForPlans: ["free"],
        message: /* @__PURE__ */ i(A, { children: [
          "Upgrade to ",
          /* @__PURE__ */ e("b", { children: "PLUS" }),
          " to customise the logo and colors"
        ] })
      }
    )
  ] });
}
const lo = () => /* @__PURE__ */ e(A, { children: /* @__PURE__ */ i(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    viewBox: "0 0 64 64",
    width: 32,
    height: 32,
    children: [
      /* @__PURE__ */ i(
        "linearGradient",
        {
          id: "95yY7w43Oj6n2vH63j6HJb",
          x1: "29.401",
          x2: "29.401",
          y1: "4.064",
          y2: "106.734",
          gradientTransform: "matrix(1 0 0 -1 0 66)",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ e("stop", { offset: "0", stopColor: "#ff5840" }),
            /* @__PURE__ */ e("stop", { offset: ".007", stopColor: "#ff5840" }),
            /* @__PURE__ */ e("stop", { offset: ".989", stopColor: "#fa528c" }),
            /* @__PURE__ */ e("stop", { offset: "1", stopColor: "#fa528c" })
          ]
        }
      ),
      /* @__PURE__ */ e(
        "path",
        {
          fill: "url(#95yY7w43Oj6n2vH63j6HJb)",
          d: "M47.46,15.5l-1.37,1.48c-1.34,1.44-3.5,1.67-5.15,0.6c-2.71-1.75-6.43-3.13-11-2.37 c-4.94,0.83-9.17,3.85-11.64, 7.97l-8.03-6.08C14.99,9.82,23.2,5,32.5,5c5,0,9.94,1.56,14.27,4.46 C48.81,10.83,49.13,13.71,47.46,15.5z"
        }
      ),
      /* @__PURE__ */ i(
        "linearGradient",
        {
          id: "95yY7w43Oj6n2vH63j6HJc",
          x1: "12.148",
          x2: "12.148",
          y1: ".872",
          y2: "47.812",
          gradientTransform: "matrix(1 0 0 -1 0 66)",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ e("stop", { offset: "0", stopColor: "#feaa53" }),
            /* @__PURE__ */ e("stop", { offset: ".612", stopColor: "#ffcd49" }),
            /* @__PURE__ */ e("stop", { offset: "1", stopColor: "#ffde44" })
          ]
        }
      ),
      /* @__PURE__ */ e(
        "path",
        {
          fill: "url(#95yY7w43Oj6n2vH63j6HJc)",
          d: "M16.01,30.91c-0.09,2.47,0.37,4.83,1.27,6.96l-8.21,6.05c-1.35-2.51-2.3-5.28-2.75-8.22 c-1.06-6.88,0.54-13.38, 3.95-18.6l8.03,6.08C16.93,25.47,16.1,28.11,16.01,30.91z"
        }
      ),
      /* @__PURE__ */ i(
        "linearGradient",
        {
          id: "95yY7w43Oj6n2vH63j6HJd",
          x1: "29.76",
          x2: "29.76",
          y1: "32.149",
          y2: "-6.939",
          gradientTransform: "matrix(1 0 0 -1 0 66)",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ e("stop", { offset: "0", stopColor: "#42d778" }),
            /* @__PURE__ */ e("stop", { offset: ".428", stopColor: "#3dca76" }),
            /* @__PURE__ */ e("stop", { offset: "1", stopColor: "#34b171" })
          ]
        }
      ),
      /* @__PURE__ */ e(
        "path",
        {
          fill: "url(#95yY7w43Oj6n2vH63j6HJd)",
          d: "M50.45,51.28c-4.55,4.07-10.61,6.57-17.36,6.71C22.91,58.2,13.66,52.53,9.07,43.92l8.21-6.05 C19.78,43.81, 25.67,48,32.5,48c3.94,0,7.52-1.28,10.33-3.44L50.45,51.28z"
        }
      ),
      /* @__PURE__ */ i(
        "linearGradient",
        {
          id: "95yY7w43Oj6n2vH63j6HJe",
          x1: "46",
          x2: "46",
          y1: "3.638",
          y2: "35.593",
          gradientTransform: "matrix(1 0 0 -1 0 66)",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ e("stop", { offset: "0", stopColor: "#155cde" }),
            /* @__PURE__ */ e("stop", { offset: ".278", stopColor: "#1f7fe5" }),
            /* @__PURE__ */ e("stop", { offset: ".569", stopColor: "#279ceb" }),
            /* @__PURE__ */ e("stop", { offset: ".82", stopColor: "#2cafef" }),
            /* @__PURE__ */ e("stop", { offset: "1", stopColor: "#2eb5f0" })
          ]
        }
      ),
      /* @__PURE__ */ e(
        "path",
        {
          fill: "url(#95yY7w43Oj6n2vH63j6HJe)",
          d: `M59,31.97c0.01,7.73-3.26,14.58-8.55,19.31l-7.62-6.72c2.1-1.61,3.77-3.71,4.84-6.15
        c0.29-0.66-0.2-1.41-0.92-1.41H37c-2.21,0-4-1.79-4-4v-2c0-2.21,1.79-4,4-4h17C56.75,27,59,29.22,59,31.97z`
        }
      )
    ]
  }
) });
function co({
  onClick: t,
  disabled: r
}) {
  return /* @__PURE__ */ e("div", { className: "m-4 w-full", children: /* @__PURE__ */ e(
    V,
    {
      className: re("w-full bg-white text-gray-900 dark:text-gray-900", r ? "" : "hover:text-white hover:dark:text-white"),
      style: {
        height: "40px",
        borderRadius: "4px",
        fontSize: "14px"
      },
      variant: "filled",
      disabled: r,
      onClick: t,
      children: /* @__PURE__ */ i(
        "div",
        {
          className: re("flex items-center justify-items-center "),
          children: [
            /* @__PURE__ */ e("div", { className: "flex flex-col items-center justify-center w-4.5 h-4.5", children: lo() }),
            /* @__PURE__ */ e(
              "div",
              {
                className: re("flex-grow pl-6 text-left"),
                children: "Sign in with Google"
              }
            )
          ]
        }
      )
    }
  ) });
}
function uo({
  fireCMSBackend: t,
  includeGoogleAdminScopes: r,
  includeLogo: o,
  includeGoogleDisclosure: n,
  includeTermsAndNewsLetter: a
}) {
  const [l, s] = x(!1), [d, p] = x(!1);
  function u() {
    let h;
    const S = ["auth/popup-closed-by-user", "auth/cancelled-popup-request"];
    return t.authProviderError && !S.includes(t.authProviderError.code) && (h = /* @__PURE__ */ e("div", { className: "p-4", children: /* @__PURE__ */ e(le, { error: t.authProviderError }) })), h;
  }
  return /* @__PURE__ */ i("div", { className: "flex flex-col items-center justify-center min-w-full p-2", children: [
    o && /* @__PURE__ */ e("div", { className: "m-4", style: {
      width: "260px",
      height: "260px"
    }, children: /* @__PURE__ */ e(qe, {}) }),
    o && /* @__PURE__ */ i(
      g,
      {
        variant: "h4",
        color: "primary",
        className: "mb-4",
        children: [
          "FireCMS ",
          /* @__PURE__ */ e(
            g,
            {
              variant: "h4",
              component: "span",
              className: "text-blue-500",
              children: "CLOUD"
            }
          )
        ]
      }
    ),
    u(),
    a && /* @__PURE__ */ i(A, { children: [
      /* @__PURE__ */ e(
        _e,
        {
          size: "small",
          invisible: !0,
          value: d,
          onValueChange: p,
          position: "start",
          label: /* @__PURE__ */ e(g, { variant: "caption", color: "primary", children: "Join our newsletter. No spam, only important updates!" })
        }
      ),
      /* @__PURE__ */ e(
        _e,
        {
          size: "small",
          invisible: !0,
          value: l,
          onValueChange: s,
          position: "start",
          label: /* @__PURE__ */ i(g, { variant: "caption", color: "primary", children: [
            "By signing in you agree to our ",
            /* @__PURE__ */ e(
              "a",
              {
                target: "_blank",
                rel: "noopener noreferrer",
                href: "https://firecms.co/policy/terms_conditions",
                children: "Terms and Conditions"
              }
            ),
            " and our ",
            /* @__PURE__ */ e(
              "a",
              {
                target: "_blank",
                rel: "noopener noreferrer",
                href: "https://firecms.co/policy/privacy_policy",
                children: "Privacy policy"
              }
            )
          ] })
        }
      )
    ] }),
    /* @__PURE__ */ e(
      co,
      {
        disabled: !l && a,
        onClick: () => {
          console.log("Google login"), t.googleLogin(r).then((h) => {
            d && h?.email && fo(h.email);
          });
        }
      }
    ),
    r && t.permissionsNotGrantedError && /* @__PURE__ */ e(
      le,
      {
        error: "You need to grant additional permissions in order to manage your Google Cloud projects"
      }
    ),
    n && /* @__PURE__ */ i(g, { variant: "caption", children: [
      "FireCMS Cloud use and transfer to any other app of information received from Google APIs will adhere to ",
      /* @__PURE__ */ e(
        "a",
        {
          target: "_blank",
          rel: "noopener noreferrer",
          href: "https://developers.google.com/terms/api-services-user-data-policy#additional_requirements_for_specific_api_scopes",
          children: "Google API Services User Data Policy"
        }
      ),
      ", including the Limited Use requirements."
    ] })
  ] });
}
const fo = (t) => {
  fetch("https://api-drplyi3b6q-ey.a.run.app/notifications/newsletter", {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify({
      email_address: t,
      source: "saas"
    })
  }).then((o) => {
    console.log("newsletter response", o);
  });
};
function ho() {
  const t = he(), { plugins: r } = Mt(), o = [];
  return r && o.push(...r.map((n, a) => /* @__PURE__ */ e(z.Fragment, { children: n.homePage?.additionalActions ?? null }, n.key)).filter(Boolean)), /* @__PURE__ */ e(
    It,
    {
      additionalActions: /* @__PURE__ */ i(A, { children: [
        " ",
        o,
        " "
      ] }),
      additionalChildrenStart: (t.collections ?? []).length > 0 ? /* @__PURE__ */ e(ce, { showForPlans: ["free"] }) : void 0,
      additionalChildrenEnd: /* @__PURE__ */ e(zt, { group: "ADMIN", children: /* @__PURE__ */ e("div", { className: "grid grid-cols-12 gap-2", children: at.map((n) => /* @__PURE__ */ e(
        "div",
        {
          className: "col-span-12 sm:col-span-6 lg:col-span-4",
          children: /* @__PURE__ */ e(
            $t,
            {
              name: n.name,
              url: n.path,
              icon: /* @__PURE__ */ e(
                Ee,
                {
                  collectionOrView: n,
                  className: "text-gray-400 dark:text-gray-600"
                }
              )
            }
          )
        },
        `nav_${n.path}`
      )) }) })
    }
  );
}
function mo({
  open: t,
  closeDialog: r,
  collection: o,
  collectionConfigController: n,
  parentCollectionIds: a,
  hasOwnTextSearchImplementation: l
}) {
  const s = Ue(), d = X(), [p, u] = x(!1), [h, S] = x(!1);
  function v() {
    return n.updateCollection({
      id: o.id,
      parentCollectionIds: a,
      collectionData: {
        textSearchEnabled: !0
      }
    });
  }
  return /* @__PURE__ */ i(
    Xe,
    {
      maxWidth: "2xl",
      open: t,
      onOpenChange: (c) => c ? void 0 : r(),
      children: [
        /* @__PURE__ */ i(Ze, { className: "flex flex-col gap-4", children: [
          /* @__PURE__ */ i(g, { variant: "h5", className: "flex flex-row gap-4 items-center", children: [
            /* @__PURE__ */ e(lr, {}),
            "Enable text search"
          ] }),
          /* @__PURE__ */ e(
            ce,
            {
              includeCTA: !1,
              showForPlans: ["free"],
              message: /* @__PURE__ */ e(A, { children: "Upgrade to PLUS to use text search" })
            }
          ),
          !l && /* @__PURE__ */ i(A, { children: [
            /* @__PURE__ */ i("div", { className: "flex flex-col gap-2 my-2", children: [
              /* @__PURE__ */ e(De, { color: "warning", children: "Local text search is not recommended for large collections." }),
              /* @__PURE__ */ e(g, { variant: "caption", children: "Note that enabling local text search will need to fetch all documents from your collection and store them in the browser. This can be inefficient for large collections. It can also incur in additional costs." }),
              /* @__PURE__ */ e(g, { variant: "caption", children: "For larger collections, you are encouraged to use an external search engine such as Algolia or Elastic Search, and assign a search delegate to your config." })
            ] }),
            /* @__PURE__ */ e(g, { children: "Local text search is the simplest way to enable text search in your collection. It loads all documents in the collection in the browser and performs the search locally. This is the recommended option for small collections." })
          ] }),
          l && /* @__PURE__ */ e(A, { children: /* @__PURE__ */ e(g, { children: "You have implemented your own text search controller. You enable text search for your collection." }) }),
          /* @__PURE__ */ i("div", { className: "flex items-end justify-end gap-4", children: [
            (l || d.localTextSearchEnabled) && !o.textSearchEnabled && d.canUseLocalTextSearch && /* @__PURE__ */ e(
              fe,
              {
                variant: "outlined",
                loading: h,
                size: "large",
                onClick: () => {
                  S(!0), v().then(() => {
                    s.open({
                      message: "Local text search enabled",
                      type: "success"
                    }), r();
                  }).finally(() => S(!1));
                },
                disabled: !d.canUseLocalTextSearch,
                children: "Enable for this collection"
              }
            ),
            !l && !d.localTextSearchEnabled && /* @__PURE__ */ e(
              fe,
              {
                variant: "outlined",
                loading: p,
                size: "large",
                onClick: () => {
                  u(!0), d.updateLocalTextSearchEnabled(!0).then(async () => {
                    o.textSearchEnabled || await v(), s.open({
                      message: "Local text search enabled",
                      type: "success"
                    }), r();
                  }).finally(() => u(!1));
                },
                disabled: !d.canUseLocalTextSearch,
                children: "Enable for project"
              }
            ),
            !d.canUseLocalTextSearch && /* @__PURE__ */ e(
              gt,
              {
                includePriceSelect: !0,
                largePriceLabel: !1,
                includePriceLabel: !1
              }
            )
          ] })
        ] }),
        /* @__PURE__ */ e(
          ke,
          {
            className: "absolute top-4 right-4",
            onClick: r,
            children: /* @__PURE__ */ e(et, {})
          }
        )
      ]
    }
  );
}
function go({ introMode: t }) {
  const r = We(), o = he(), n = ot(), a = n.configPermissions ? n.configPermissions({
    user: r.user
  }).createCollections : !0, [l, s] = z.useState(void 0);
  if (k(() => {
    n.getPathSuggestions?.("").then(s);
  }, [n.getPathSuggestions]), !n.getPathSuggestions)
    return null;
  const d = (l ?? []).length > 3 || (o.collections ?? []).length === 0 && (l ?? []).length > 0;
  return /* @__PURE__ */ e(
    cr,
    {
      in: t === "existing_project" || d,
      children: /* @__PURE__ */ i(
        "div",
        {
          className: "flex flex-col gap-1 p-2 my-4",
          children: [
            !t && /* @__PURE__ */ i(g, { variant: "body2", color: "secondary", children: [
              "Create a collection ",
              /* @__PURE__ */ e("b", { children: "automatically" }),
              " from your data:"
            ] }),
            t === "existing_project" && /* @__PURE__ */ i(g, { children: [
              "You will see your ",
              /* @__PURE__ */ e("b", { children: "database collections" }),
              " here, a few seconds after project creation"
            ] }),
            /* @__PURE__ */ i(
              "div",
              {
                className: "flex flex-row gap-1 overflow-scroll no-scrollbar ",
                children: [
                  (l ?? []).map((u) => /* @__PURE__ */ e("div", { children: /* @__PURE__ */ e(
                    K,
                    {
                      icon: /* @__PURE__ */ e(Qe, { size: "small" }),
                      colorScheme: "cyanLighter",
                      onClick: n && a ? () => n.createCollection({
                        initialValues: {
                          path: u,
                          name: Vt(u)
                        },
                        parentCollectionIds: [],
                        redirect: !0,
                        sourceClick: "root_collection_suggestion"
                      }) : void 0,
                      size: "small",
                      children: u
                    }
                  ) }, u)),
                  l === void 0 && /* @__PURE__ */ e(tt, { size: "small" }),
                  l?.length === 0 && /* @__PURE__ */ e(g, { variant: "caption", children: "No suggestions" })
                ]
              }
            )
          ]
        }
      )
    }
  );
}
function po({
  projectConfig: t,
  firestoreDelegate: r,
  collectionConfigController: o,
  appConfig: n,
  introMode: a
}) {
  const l = !!n?.textSearchControllerBuilder;
  return {
    key: "saas",
    homePage: {
      additionalChildrenStart: a ? /* @__PURE__ */ e(yo, { introMode: a }) : void 0,
      additionalChildrenEnd: /* @__PURE__ */ e(go, { introMode: a })
    },
    collectionView: {
      showTextSearchBar: ({
        context: s,
        path: d,
        collection: p
      }) => p.textSearchEnabled !== !1,
      onTextSearchClick: ({
        context: s,
        path: d,
        collection: p,
        parentCollectionIds: u
      }) => t.canUseLocalTextSearch && t.localTextSearchEnabled && p.textSearchEnabled ? r.initTextSearchController({
        path: d,
        collection: p
      }) : (u === void 0 ? console.warn("Enabling text search: Parent collection ids are undefined") : s.dialogsController.open({
        key: "text_search_info",
        Component: (h) => /* @__PURE__ */ e(
          mo,
          {
            ...h,
            hasOwnTextSearchImplementation: l,
            collectionConfigController: o,
            parentCollectionIds: u,
            path: d,
            collection: p
          }
        )
      }), Promise.resolve(!1))
    }
  };
}
function yo({ introMode: t }) {
  if (!he().topLevelNavigation)
    throw Error("Navigation not ready in FireCMSHomePage");
  return /* @__PURE__ */ i("div", { className: "mt-8 flex flex-col p-2", children: [
    /* @__PURE__ */ e(g, { variant: "h4", className: "mb-4", children: "Welcome" }),
    /* @__PURE__ */ e(g, { paragraph: !0, children: "Your admin panel is ready âœŒï¸" }),
    /* @__PURE__ */ e(g, { paragraph: !0, children: "FireCMS can be used as a standalone admin panel but it shines when you add your own custom functionality. Including your own custom components, fields, actions, views, and more." }),
    /* @__PURE__ */ i("div", { children: [
      /* @__PURE__ */ e(g, { className: "inline", children: "Get started with:" }),
      /* @__PURE__ */ e(
        "div",
        {
          className: "ml-2 select-all font-mono bg-gray-100 text-gray-800 dark:bg-gray-800 dark:text-white p-2 px-3  border-gray-200 border-solid w-fit text-md font-bold inline-flex rounded-md",
          children: "yarn create firecms-app"
        }
      ),
      /* @__PURE__ */ i(g, { className: "inline ml-2", children: [
        "More info in the ",
        /* @__PURE__ */ e(
          "a",
          {
            href: "https://firecms.co/docs/customization_quickstart",
            rel: "noopener noreferrer",
            target: "_blank",
            children: "docs"
          }
        )
      ] })
    ] })
  ] });
}
const Je = 200;
function vo({
  projectId: t,
  appConfig: r,
  backendApiHost: o = "https://api-drplyi3b6q-ey.a.run.app",
  // TODO
  onAnalyticsEvent: n,
  basePath: a,
  baseCollectionPath: l
}) {
  const s = Bt(), {
    firebaseApp: d,
    firebaseConfigLoading: p,
    configError: u,
    firebaseConfigError: h
  } = it({
    name: "firecms-backend",
    fromUrl: o + "/config"
  }), S = Mr({
    backendApiHost: o,
    backendFirebaseApp: d
  });
  let v;
  return p || !d ? v = /* @__PURE__ */ e(Fe, { projectId: t, text: "Backend loading" }) : h ? v = /* @__PURE__ */ e(
    le,
    {
      error: h
    }
  ) : u ? v = /* @__PURE__ */ e(
    le,
    {
      error: u
    }
  ) : S.authLoading ? v = /* @__PURE__ */ e(Fe, { projectId: t, text: "Auth loading" }) : S.user ? v = /* @__PURE__ */ e(
    xo,
    {
      fireCMSBackend: S,
      projectId: t,
      modeController: s,
      appConfig: r,
      basePath: a,
      baseCollectionPath: l,
      onAnalyticsEvent: n
    }
  ) : v = /* @__PURE__ */ e(de, { maxWidth: "lg", children: /* @__PURE__ */ e(
    uo,
    {
      fireCMSBackend: S,
      includeLogo: !0,
      includeGoogleAdminScopes: !1,
      includeTermsAndNewsLetter: !1,
      includeGoogleDisclosure: !1
    }
  ) }), /* @__PURE__ */ e(hr, { basename: a, children: v });
}
function Fe(t) {
  return /* @__PURE__ */ e(
    ye,
    {
      name: t.projectConfig?.projectName ?? "",
      logo: t.projectConfig?.logo,
      FireCMSAppBar: t.FireCMSAppBarComponent,
      includeDrawer: !1,
      children: /* @__PURE__ */ e(q, { text: t.text })
    },
    "project_scaffold_" + t.projectId
  );
}
const xo = function({
  projectId: r,
  fireCMSBackend: o,
  ...n
}) {
  const a = Fr({
    projectId: r,
    backendFirebaseApp: o.backendFirebaseApp
  }), l = zr({
    backendFirebaseApp: o.backendFirebaseApp,
    projectId: r,
    projectsApi: o.projectsApi,
    usersLimit: a.usersLimit,
    canEditRoles: a.canEditRoles,
    fireCMSBackend: o
  });
  return l.loading || !a.clientFirebaseConfig && !a.configError ? /* @__PURE__ */ e(
    Fe,
    {
      projectId: r,
      projectConfig: a,
      FireCMSAppBarComponent: n.FireCMSAppBarComponent,
      text: "Client loading"
    }
  ) : /* @__PURE__ */ e(
    bo,
    {
      projectId: r,
      projectConfig: a,
      fireCMSBackend: o,
      customizationLoading: !1,
      userManagement: l,
      ...n
    }
  );
};
function Co({
  configError: t,
  onLogout: r,
  fireCMSBackend: o
}) {
  const n = "code" in t ? /* @__PURE__ */ e(
    Yr,
    {
      error: t,
      fireCMSBackend: o
    }
  ) : /* @__PURE__ */ i(A, { children: [
    /* @__PURE__ */ e(g, { children: t.message }),
    /* @__PURE__ */ i(g, { children: [
      "This error may be caused when a user has been deleted from the client project. Make sure a user exists in the client project with the same email as the one trying to log in. If the problem persists, reach us at ",
      /* @__PURE__ */ e(
        "a",
        {
          href: "mailto:hello@firecms.co?subject=FireCMS%20login%20error",
          rel: "noopener noreferrer",
          target: "_blank",
          children: " hello@firecms.co "
        }
      ),
      ", or in our ",
      /* @__PURE__ */ e(
        "a",
        {
          rel: "noopener noreferrer",
          target: "_blank",
          href: "https://discord.gg/fxy7xsQm3m",
          children: "Discord channel"
        }
      ),
      "."
    ] })
  ] });
  return /* @__PURE__ */ i(de, { maxWidth: "2xl", className: "flex flex-col gap-4", children: [
    /* @__PURE__ */ i("div", { className: "flex gap-4 items-center", children: [
      /* @__PURE__ */ e(rt, { color: "error" }),
      /* @__PURE__ */ e(g, { variant: "h4", children: "Error logging in" })
    ] }),
    n,
    /* @__PURE__ */ e(V, { variant: "outlined", onClick: r, children: "Sign out" })
  ] });
}
function wo(t) {
  return /* @__PURE__ */ i(de, { maxWidth: "2xl", className: "flex flex-col gap-4", children: [
    /* @__PURE__ */ i("div", { className: "flex gap-4 items-center", children: [
      /* @__PURE__ */ e(rt, { color: "error" }),
      /* @__PURE__ */ e(g, { variant: "h4", children: "Error accessing project" })
    ] }),
    /* @__PURE__ */ e(g, { children: t.configError.message }),
    /* @__PURE__ */ e(g, { children: "This error may be caused when trying to access with a user that is not registered in the project. You can ask the project owner to add you to the project." }),
    /* @__PURE__ */ e(V, { variant: "outlined", onClick: t.onLogout, children: "Sign out" })
  ] });
}
function bo({
  projectConfig: t,
  userManagement: r,
  projectId: o,
  fireCMSBackend: n,
  appConfig: a,
  customizationLoading: l,
  ...s
}) {
  const [d, p] = x(), {
    firebaseApp: u,
    firebaseConfigLoading: h,
    configError: S
  } = it({
    onFirebaseInit: a?.onFirebaseInit,
    firebaseConfig: t.clientFirebaseConfig
  }), v = pr({
    firebaseApp: u,
    options: a?.appCheck
  }), c = yr({
    firebaseApp: u,
    onSignOut: n.signOut,
    defineRolesFor: r.defineRolesFor
  }), m = Pe(
    () => {
      if (r.loading || c.authLoading)
        return;
      const P = c.user;
      if (P)
        return r.users.find((F) => F.email.toLowerCase() === P?.email?.toLowerCase());
    },
    [c.authLoading, c.user, r.loading, r.users]
  ), {
    delegatedLoginLoading: y,
    delegatedLoginError: f
  } = kr({
    projectsApi: n.projectsApi,
    firebaseApp: u,
    projectId: o,
    onUserChanged: (P) => {
      c.setUser(P ?? null), c.setRoles(m?.roles ?? []);
    },
    onAnalyticsEvent: s.onAnalyticsEvent
  }), E = j(({
    pathSegments: P,
    collection: F,
    user: O,
    entity: U
  }) => Ge({
    collection: F,
    user: O
  }), []), T = Sr({
    firebaseApp: n.backendFirebaseApp,
    configPath: `projects/${o}/collections`,
    permissions: E,
    propertyConfigs: a?.propertyConfigs
  });
  let C;
  return r.loading ? C = /* @__PURE__ */ e(q, { text: "Project loading" }) : v.loading ? C = /* @__PURE__ */ e(q, { text: "AppCheck loading" }) : f ? (console.log("Delegated login error", f), C = /* @__PURE__ */ e(
    Co,
    {
      configError: f,
      fireCMSBackend: n,
      onLogout: n.signOut
    }
  )) : d ? (console.warn("No user was found with email " + d.email), C = /* @__PURE__ */ e(Ye, { authController: c })) : t.configError ? C = /* @__PURE__ */ e(
    wo,
    {
      configError: t.configError,
      onLogout: n.signOut
    }
  ) : l ? C = /* @__PURE__ */ e(q, { text: "Project customization loading" }) : h ? C = /* @__PURE__ */ e(q, { text: "Client config loading" }) : S ? C = /* @__PURE__ */ e(de, { children: /* @__PURE__ */ e(le, { error: S ?? "Error fetching client Firebase config" }) }) : y ? C = /* @__PURE__ */ e(q, { text: "Logging in to " + o }) : c.user ? m || (C = /* @__PURE__ */ e(Ye, { authController: c })) : C = /* @__PURE__ */ e(q, { text: "Auth loading" }), C ? /* @__PURE__ */ e(
    ye,
    {
      name: t.projectName ?? "",
      logo: t.logo,
      includeDrawer: !1,
      FireCMSAppBar: s.FireCMSAppBarComponent,
      children: C
    },
    "project_scaffold_" + t.projectId
  ) : /* @__PURE__ */ e(
    Po,
    {
      fireCMSUser: m,
      fireCMSBackend: n,
      appConfig: a,
      authController: c,
      projectConfig: t,
      collectionConfigController: T,
      firebaseApp: u,
      userManagement: r,
      ...s
    }
  );
}
function Ye({ authController: t }) {
  return /* @__PURE__ */ i(de, { maxWidth: "md", className: "gap-4", children: [
    /* @__PURE__ */ e(
      le,
      {
        title: "You don't have access to this project",
        error: "You can request permission to the owner"
      }
    ),
    /* @__PURE__ */ e(V, { variant: "text", onClick: t.signOut, children: "Sign out" })
  ] });
}
function So(t, r, o) {
  const {
    collections: n
  } = o, a = (n ?? []).map((p) => p.path.trim().toLowerCase()), [l, s] = z.useState();
  return k(() => {
    const p = t.googleCredential?.accessToken;
    t.projectsApi.getRootCollections(r.projectId, p).then((u) => {
      s(u.filter((h) => !a.includes(h.trim().toLowerCase())));
    });
  }, []), { getPathSuggestions: j((p) => !p && l ? Promise.resolve(l) : Promise.resolve([]), [l]) };
}
function Po({
  fireCMSUser: t,
  firebaseApp: r,
  projectConfig: o,
  userManagement: n,
  collectionConfigController: a,
  appConfig: l,
  authController: s,
  modeController: d,
  fireCMSBackend: p,
  FireCMSAppBarComponent: u,
  onAnalyticsEvent: h,
  basePath: S,
  baseCollectionPath: v
}) {
  if (!s.user)
    throw Error("You can only use FireCMSAppAuthenticated with an authenticated user");
  const c = Pe(() => No(n.usersLimit), [n.usersLimit]), m = j(({
    user: N,
    collection: R
  }) => Pr({
    user: t,
    userManagement: n,
    collection: R
  }), [n, t]), y = Pe(() => {
    const N = {};
    return l?.propertyConfigs?.forEach((R) => {
      N[R.key] = R;
    }), N;
  }, [l?.propertyConfigs]), f = br({
    exportAllowed: j(({ collectionEntitiesCount: N }) => o.canExport || N <= Je, [o.canExport]),
    onAnalyticsEvent: h,
    notAllowedView: /* @__PURE__ */ e(
      ce,
      {
        showForPlans: ["free"],
        message: `Upgrade to export more than ${Je} entities`
      }
    )
  }), E = gr({
    SubscriptionMessage: Hr,
    host: p.backendApiHost
  });
  je(o.projectName ?? "", o.logo);
  const T = Ht(), C = vr({
    firebaseApp: r,
    textSearchControllerBuilder: l?.textSearchControllerBuilder,
    firestoreIndexesBuilder: l?.firestoreIndexesBuilder,
    localTextSearchEnabled: o.localTextSearchEnabled
  }), P = xr({
    firebaseApp: r
  }), F = Jt({
    basePath: S,
    baseCollectionPath: v,
    authController: s,
    collections: l?.collections,
    views: l?.views,
    userConfigPersistence: T,
    dataSourceDelegate: C,
    injectCollections: j(
      (N) => dr(
        N,
        a.collections ?? [],
        l?.modifyCollection
      ),
      [l?.modifyCollection, a.collections]
    )
  }), O = F.initialised && F.collections !== void 0 && F.collections.length === 0, U = po({
    projectConfig: o,
    firestoreDelegate: C,
    collectionConfigController: a,
    appConfig: l,
    introMode: O ? o.creationType === "new" ? "new_project" : "existing_project" : void 0
  }), { getPathSuggestions: D } = So(p, o, F), $ = ur({
    collectionConfigController: a,
    configPermissions: m,
    reservedGroups: st,
    getPathSuggestions: D,
    getUser: (N) => n.users.find((R) => R.uid === N) ?? null,
    collectionInference: Cr(r),
    getData: (N, R) => wr(r, N, R, 400),
    onAnalyticsEvent: h
  }), B = [U, f, $, E];
  return /* @__PURE__ */ e(Rr, { ...p, children: /* @__PURE__ */ e(jr, { config: o, children: /* @__PURE__ */ e(Tt, { userManagement: n, children: /* @__PURE__ */ e(Yt, { children: /* @__PURE__ */ e(Gt, { value: d, children: /* @__PURE__ */ e(
    Wt,
    {
      navigationController: F,
      dateTimeFormat: l?.dateTimeFormat,
      entityViews: l?.entityViews,
      locale: l?.locale,
      propertyConfigs: y,
      authController: s,
      userConfigPersistence: T,
      dataSourceDelegate: C,
      storageSource: P,
      entityLinkBuilder: ({ entity: N }) => `https://console.firebase.google.com/project/${r.options.projectId}/firestore/data/${N.path}/${N.id}`,
      onAnalyticsEvent: h,
      plugins: B,
      components: {
        missingReference: fr
      },
      children: ({
        context: N,
        loading: R
      }) => {
        let b;
        return R ? b = /* @__PURE__ */ e(
          ye,
          {
            name: o.projectName ?? "",
            logo: o.logo,
            includeDrawer: !1,
            FireCMSAppBar: u,
            fireCMSAppBarProps: l?.fireCMSAppBarComponentProps,
            children: /* @__PURE__ */ e(q, { text: "Almost there" })
          },
          "project_scaffold_" + o.projectId
        ) : b = /* @__PURE__ */ i(
          ye,
          {
            name: o.projectName ?? "",
            logo: o.logo,
            Drawer: Br,
            FireCMSAppBar: u,
            fireCMSAppBarProps: l?.fireCMSAppBarComponentProps,
            autoOpenDrawer: l?.autoOpenDrawer,
            children: [
              /* @__PURE__ */ e(
                qt,
                {
                  HomePage: l?.HomePage ?? ho,
                  customRoutes: c
                }
              ),
              /* @__PURE__ */ e(Qt, {})
            ]
          },
          "project_scaffold_" + o.projectId
        ), b;
      }
    }
  ) }) }) }) }) });
}
function No(t) {
  return [
    {
      path: "users",
      name: "CMS Users",
      group: "Admin",
      icon: "face",
      hideFromNavigation: !0,
      view: /* @__PURE__ */ e(Ft, { children: /* @__PURE__ */ e(
        ce,
        {
          showForPlans: ["free"],
          message: /* @__PURE__ */ i(A, { children: [
            "Upgrade to PLUS to remove the ",
            /* @__PURE__ */ i("b", { children: [
              t,
              " users limit"
            ] })
          ] })
        }
      ) })
    },
    {
      path: "roles",
      name: "Roles",
      group: "Admin",
      icon: "gpp_good",
      hideFromNavigation: !0,
      view: /* @__PURE__ */ e(Ut, { children: /* @__PURE__ */ e(
        ce,
        {
          showForPlans: ["free"],
          message: /* @__PURE__ */ i(A, { children: [
            "Upgrade to PLUS to be able to customise ",
            /* @__PURE__ */ e("b", { children: "roles" })
          ] })
        }
      ) })
    },
    {
      path: "settings",
      name: "Project settings",
      group: "Admin",
      icon: "settings",
      hideFromNavigation: !0,
      view: /* @__PURE__ */ e(oo, {})
    }
  ].map(({
    path: r,
    name: o,
    view: n
  }) => /* @__PURE__ */ e(
    mr,
    {
      path: r,
      element: n
    },
    "navigation_admin_" + r
  ));
}
function Vo(t) {
  return k(() => {
    console.warn("IMPORTANT! The FireCMSApp component is deprecated and will be removed in future versions. Please use FireCMSCloudApp instead.");
  }, []), /* @__PURE__ */ e(vo, { ...t });
}
export {
  at as ADMIN_VIEWS_CONFIG,
  $o as ActiveSubscriptions,
  Le as ApiError,
  Dr as CHECKOUT_SESSION_COLLECTION,
  Be as CUSTOMERS_COLLECTION,
  Yr as CloudErrorView,
  ro as CurrentSubscriptionView,
  Vo as FireCMSApp,
  Rr as FireCMSBackEndProvider,
  ct as FireCMSBackendContext,
  xo as FireCMSClient,
  bo as FireCMSClientWithController,
  vo as FireCMSCloudApp,
  Br as FireCMSCloudDrawer,
  Hr as FireCMSDataEnhancementSubscriptionMessage,
  uo as FireCMSLoginView,
  co as GoogleLoginButton,
  Ve as PRODUCTS_COLLECTION,
  ht as PlanChip,
  Xr as PlansComparison,
  Kr as PlansComparisonDialog,
  mt as ProductView,
  lt as ProjectConfigContext,
  jr as ProjectConfigProvider,
  oo as ProjectSettings,
  eo as ProjectSubscriptionPlans,
  st as RESERVED_GROUPS,
  Ar as SUBSCRIPTIONS_COLLECTION,
  Jr as SecurityRulesInstructions,
  Zr as StripeDisclaimer,
  ce as SubscriptionPlanWidget,
  zo as areRolesEqual,
  Or as buildProjectsApi,
  Er as cacheDelegatedLoginToken,
  Lr as clearDelegatedLoginTokensCache,
  be as darkenColor,
  _r as getDelegatedLoginTokenFromCache,
  $e as hexToRgbaWithOpacity,
  Pr as resolveCollectionConfigPermissions,
  zr as useBuildCloudUserManagement,
  Mr as useBuildFireCMSBackend,
  Fr as useBuildProjectConfig,
  kr as useDelegatedLogin,
  oe as useFireCMSBackend,
  X as useProjectConfig,
  dt as useSubscriptionsForUserController
};
